Changes in module 'nf-core/fastqc'
--- modules/nf-core/fastqc/main.nf
+++ modules/nf-core/fastqc/main.nf
@@ -5,7 +5,7 @@
     conda "bioconda::fastqc=0.11.9"
     container "${ workflow.containerEngine == 'singularity' && !task.ext.singularity_pull_docker_container ?
         'https://depot.galaxyproject.org/singularity/fastqc:0.11.9--0' :
-        'biocontainers/fastqc:0.11.9--0' }"
+        'quay.io/biocontainers/fastqc:0.11.9--0' }"
 
     input:
     tuple val(meta), path(reads)

************************************************************
