
<!doctype html>
<html lang="en" class="no-js">
  <head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      
      
      
        <link rel="canonical" href="https://genomic-medicine-sweden.github.io/nallo/dev/output/">
      
      
        <link rel="prev" href="../usage/">
      
      
        <link rel="next" href="../parameters/">
      
      
      <link rel="icon" href="../assets/images/favicon.png">
      <meta name="generator" content="mkdocs-1.6.1, mkdocs-material-9.5.41">
    
    
      
        <title>Output - genomic-medicine-sweden/nallo</title>
      
    
    
      <link rel="stylesheet" href="../assets/stylesheets/main.0253249f.min.css">
      
        
        <link rel="stylesheet" href="../assets/stylesheets/palette.06af60db.min.css">
      
      


    
    
      
    
    
      
        
        
        <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Inter:300,300i,400,400i,700,700i%7CRoboto+Mono:400,400i,700,700i&display=fallback">
        <style>:root{--md-text-font:"Inter";--md-code-font:"Roboto Mono"}</style>
      
    
    
    <script>__md_scope=new URL("..",location),__md_hash=e=>[...e].reduce(((e,_)=>(e<<5)-e+_.charCodeAt(0)),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script>
    
      

    
    
    
  </head>
  
  
    
    
      
    
    
    
    
    <body dir="ltr" data-md-color-scheme="slate" data-md-color-primary="custom" data-md-color-accent="custom">
  
    
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off">
    <label class="md-overlay" for="__drawer"></label>
    <div data-md-component="skip">
      
        
        <a href="#genomic-medicine-swedennallo-output" class="md-skip">
          Skip to content
        </a>
      
    </div>
    <div data-md-component="announce">
      
    </div>
    
      <div data-md-color-scheme="default" data-md-component="outdated" hidden>
        
      </div>
    
    
      

  

<header class="md-header md-header--shadow" data-md-component="header">
  <nav class="md-header__inner md-grid" aria-label="Header">
    <a href=".." title="genomic-medicine-sweden/nallo" class="md-header__button md-logo" aria-label="genomic-medicine-sweden/nallo" data-md-component="logo">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 16 16"><path d="M2.5 1.75v11.5c0 .138.112.25.25.25h3.17a.75.75 0 0 1 0 1.5H2.75A1.75 1.75 0 0 1 1 13.25V1.75C1 .784 1.784 0 2.75 0h8.5C12.216 0 13 .784 13 1.75v7.736a.75.75 0 0 1-1.5 0V1.75a.25.25 0 0 0-.25-.25h-8.5a.25.25 0 0 0-.25.25m13.274 9.537zl-4.557 4.45a.75.75 0 0 1-1.055-.008l-1.943-1.95a.75.75 0 0 1 1.062-1.058l1.419 1.425 4.026-3.932a.75.75 0 1 1 1.048 1.074M4.75 4h4.5a.75.75 0 0 1 0 1.5h-4.5a.75.75 0 0 1 0-1.5M4 7.75A.75.75 0 0 1 4.75 7h2a.75.75 0 0 1 0 1.5h-2A.75.75 0 0 1 4 7.75"/></svg>

    </a>
    <label class="md-header__button md-icon" for="__drawer">
      
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M3 6h18v2H3zm0 5h18v2H3zm0 5h18v2H3z"/></svg>
    </label>
    <div class="md-header__title" data-md-component="header-title">
      <div class="md-header__ellipsis">
        <div class="md-header__topic">
          <span class="md-ellipsis">
            genomic-medicine-sweden/nallo
          </span>
        </div>
        <div class="md-header__topic" data-md-component="header-topic">
          <span class="md-ellipsis">
            
              Output
            
          </span>
        </div>
      </div>
    </div>
    
      
        <form class="md-header__option" data-md-component="palette">
  
    
    
    
    <input class="md-option" data-md-color-media="(prefers-color-scheme: dark)" data-md-color-scheme="slate" data-md-color-primary="custom" data-md-color-accent="custom"  aria-label="Switch to light mode"  type="radio" name="__palette" id="__palette_0">
    
      <label class="md-header__button md-icon" title="Switch to light mode" for="__palette_1" hidden>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 7a5 5 0 0 1 5 5 5 5 0 0 1-5 5 5 5 0 0 1-5-5 5 5 0 0 1 5-5m0 2a3 3 0 0 0-3 3 3 3 0 0 0 3 3 3 3 0 0 0 3-3 3 3 0 0 0-3-3m0-7 2.39 3.42C13.65 5.15 12.84 5 12 5s-1.65.15-2.39.42zM3.34 7l4.16-.35A7.2 7.2 0 0 0 5.94 8.5c-.44.74-.69 1.5-.83 2.29zm.02 10 1.76-3.77a7.131 7.131 0 0 0 2.38 4.14zM20.65 7l-1.77 3.79a7.02 7.02 0 0 0-2.38-4.15zm-.01 10-4.14.36c.59-.51 1.12-1.14 1.54-1.86.42-.73.69-1.5.83-2.29zM12 22l-2.41-3.44c.74.27 1.55.44 2.41.44.82 0 1.63-.17 2.37-.44z"/></svg>
      </label>
    
  
    
    
    
    <input class="md-option" data-md-color-media="(prefers-color-scheme: light)" data-md-color-scheme="default" data-md-color-primary="custom" data-md-color-accent="custom"  aria-label="Switch to dark mode"  type="radio" name="__palette" id="__palette_1">
    
      <label class="md-header__button md-icon" title="Switch to dark mode" for="__palette_0" hidden>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="m17.75 4.09-2.53 1.94.91 3.06-2.63-1.81-2.63 1.81.91-3.06-2.53-1.94L12.44 4l1.06-3 1.06 3zm3.5 6.91-1.64 1.25.59 1.98-1.7-1.17-1.7 1.17.59-1.98L15.75 11l2.06-.05L18.5 9l.69 1.95zm-2.28 4.95c.83-.08 1.72 1.1 1.19 1.85-.32.45-.66.87-1.08 1.27C15.17 23 8.84 23 4.94 19.07c-3.91-3.9-3.91-10.24 0-14.14.4-.4.82-.76 1.27-1.08.75-.53 1.93.36 1.85 1.19-.27 2.86.69 5.83 2.89 8.02a9.96 9.96 0 0 0 8.02 2.89m-1.64 2.02a12.08 12.08 0 0 1-7.8-3.47c-2.17-2.19-3.33-5-3.49-7.82-2.81 3.14-2.7 7.96.31 10.98 3.02 3.01 7.84 3.12 10.98.31"/></svg>
      </label>
    
  
</form>
      
    
    
      <script>var palette=__md_get("__palette");if(palette&&palette.color){if("(prefers-color-scheme)"===palette.color.media){var media=matchMedia("(prefers-color-scheme: light)"),input=document.querySelector(media.matches?"[data-md-color-media='(prefers-color-scheme: light)']":"[data-md-color-media='(prefers-color-scheme: dark)']");palette.color.media=input.getAttribute("data-md-color-media"),palette.color.scheme=input.getAttribute("data-md-color-scheme"),palette.color.primary=input.getAttribute("data-md-color-primary"),palette.color.accent=input.getAttribute("data-md-color-accent")}for(var[key,value]of Object.entries(palette.color))document.body.setAttribute("data-md-color-"+key,value)}</script>
    
    
    
      <label class="md-header__button md-icon" for="__search">
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"/></svg>
      </label>
      <div class="md-search" data-md-component="search" role="dialog">
  <label class="md-search__overlay" for="__search"></label>
  <div class="md-search__inner" role="search">
    <form class="md-search__form" name="search">
      <input type="text" class="md-search__input" name="query" aria-label="Search" placeholder="Search" autocapitalize="off" autocorrect="off" autocomplete="off" spellcheck="false" data-md-component="search-query" required>
      <label class="md-search__icon md-icon" for="__search">
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"/></svg>
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11z"/></svg>
      </label>
      <nav class="md-search__options" aria-label="Search">
        
          <a href="javascript:void(0)" class="md-search__icon md-icon" title="Share" aria-label="Share" data-clipboard data-clipboard-text="" data-md-component="search-share" tabindex="-1">
            
            <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M18 16.08c-.76 0-1.44.3-1.96.77L8.91 12.7c.05-.23.09-.46.09-.7s-.04-.47-.09-.7l7.05-4.11c.54.5 1.25.81 2.04.81a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3c0 .24.04.47.09.7L8.04 9.81C7.5 9.31 6.79 9 6 9a3 3 0 0 0-3 3 3 3 0 0 0 3 3c.79 0 1.5-.31 2.04-.81l7.12 4.15c-.05.21-.08.43-.08.66 0 1.61 1.31 2.91 2.92 2.91s2.92-1.3 2.92-2.91A2.92 2.92 0 0 0 18 16.08"/></svg>
          </a>
        
        <button type="reset" class="md-search__icon md-icon" title="Clear" aria-label="Clear" tabindex="-1">
          
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"/></svg>
        </button>
      </nav>
      
    </form>
    <div class="md-search__output">
      <div class="md-search__scrollwrap" tabindex="0" data-md-scrollfix>
        <div class="md-search-result" data-md-component="search-result">
          <div class="md-search-result__meta">
            Initializing search
          </div>
          <ol class="md-search-result__list" role="presentation"></ol>
        </div>
      </div>
    </div>
  </div>
</div>
    
    
      <div class="md-header__source">
        <a href="https://github.com/genomic-medicine-sweden/nallo" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 496 512"><!--! Font Awesome Free 6.6.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2024 Fonticons, Inc.--><path d="M165.9 397.4c0 2-2.3 3.6-5.2 3.6-3.3.3-5.6-1.3-5.6-3.6 0-2 2.3-3.6 5.2-3.6 3-.3 5.6 1.3 5.6 3.6m-31.1-4.5c-.7 2 1.3 4.3 4.3 4.9 2.6 1 5.6 0 6.2-2s-1.3-4.3-4.3-5.2c-2.6-.7-5.5.3-6.2 2.3m44.2-1.7c-2.9.7-4.9 2.6-4.6 4.9.3 2 2.9 3.3 5.9 2.6 2.9-.7 4.9-2.6 4.6-4.6-.3-1.9-3-3.2-5.9-2.9M244.8 8C106.1 8 0 113.3 0 252c0 110.9 69.8 205.8 169.5 239.2 12.8 2.3 17.3-5.6 17.3-12.1 0-6.2-.3-40.4-.3-61.4 0 0-70 15-84.7-29.8 0 0-11.4-29.1-27.8-36.6 0 0-22.9-15.7 1.6-15.4 0 0 24.9 2 38.6 25.8 21.9 38.6 58.6 27.5 72.9 20.9 2.3-16 8.8-27.1 16-33.7-55.9-6.2-112.3-14.3-112.3-110.5 0-27.5 7.6-41.3 23.6-58.9-2.6-6.5-11.1-33.3 2.6-67.9 20.9-6.5 69 27 69 27 20-5.6 41.5-8.5 62.8-8.5s42.8 2.9 62.8 8.5c0 0 48.1-33.6 69-27 13.7 34.7 5.2 61.4 2.6 67.9 16 17.7 25.8 31.5 25.8 58.9 0 96.5-58.9 104.2-114.8 110.5 9.2 7.9 17 22.9 17 46.4 0 33.7-.3 75.4-.3 83.6 0 6.5 4.6 14.4 17.3 12.1C428.2 457.8 496 362.9 496 252 496 113.3 383.5 8 244.8 8M97.2 352.9c-1.3 1-1 3.3.7 5.2 1.6 1.6 3.9 2.3 5.2 1 1.3-1 1-3.3-.7-5.2-1.6-1.6-3.9-2.3-5.2-1m-10.8-8.1c-.7 1.3.3 2.9 2.3 3.9 1.6 1 3.6.7 4.3-.7.7-1.3-.3-2.9-2.3-3.9-2-.6-3.6-.3-4.3.7m32.4 35.6c-1.6 1.3-1 4.3 1.3 6.2 2.3 2.3 5.2 2.6 6.5 1 1.3-1.3.7-4.3-1.3-6.2-2.2-2.3-5.2-2.6-6.5-1m-11.4-14.7c-1.6 1-1.6 3.6 0 5.9s4.3 3.3 5.6 2.3c1.6-1.3 1.6-3.9 0-6.2-1.4-2.3-4-3.3-5.6-2"/></svg>
  </div>
  <div class="md-source__repository">
    genomic-medicine-sweden/nallo
  </div>
</a>
      </div>
    
  </nav>
  
</header>
    
    <div class="md-container" data-md-component="container">
      
      
        
          
        
      
      <main class="md-main" data-md-component="main">
        <div class="md-main__inner md-grid">
          
            
              
              <div class="md-sidebar md-sidebar--primary" data-md-component="sidebar" data-md-type="navigation" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    



<nav class="md-nav md-nav--primary" aria-label="Navigation" data-md-level="0">
  <label class="md-nav__title" for="__drawer">
    <a href=".." title="genomic-medicine-sweden/nallo" class="md-nav__button md-logo" aria-label="genomic-medicine-sweden/nallo" data-md-component="logo">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 16 16"><path d="M2.5 1.75v11.5c0 .138.112.25.25.25h3.17a.75.75 0 0 1 0 1.5H2.75A1.75 1.75 0 0 1 1 13.25V1.75C1 .784 1.784 0 2.75 0h8.5C12.216 0 13 .784 13 1.75v7.736a.75.75 0 0 1-1.5 0V1.75a.25.25 0 0 0-.25-.25h-8.5a.25.25 0 0 0-.25.25m13.274 9.537zl-4.557 4.45a.75.75 0 0 1-1.055-.008l-1.943-1.95a.75.75 0 0 1 1.062-1.058l1.419 1.425 4.026-3.932a.75.75 0 1 1 1.048 1.074M4.75 4h4.5a.75.75 0 0 1 0 1.5h-4.5a.75.75 0 0 1 0-1.5M4 7.75A.75.75 0 0 1 4.75 7h2a.75.75 0 0 1 0 1.5h-2A.75.75 0 0 1 4 7.75"/></svg>

    </a>
    genomic-medicine-sweden/nallo
  </label>
  
    <div class="md-nav__source">
      <a href="https://github.com/genomic-medicine-sweden/nallo" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 496 512"><!--! Font Awesome Free 6.6.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2024 Fonticons, Inc.--><path d="M165.9 397.4c0 2-2.3 3.6-5.2 3.6-3.3.3-5.6-1.3-5.6-3.6 0-2 2.3-3.6 5.2-3.6 3-.3 5.6 1.3 5.6 3.6m-31.1-4.5c-.7 2 1.3 4.3 4.3 4.9 2.6 1 5.6 0 6.2-2s-1.3-4.3-4.3-5.2c-2.6-.7-5.5.3-6.2 2.3m44.2-1.7c-2.9.7-4.9 2.6-4.6 4.9.3 2 2.9 3.3 5.9 2.6 2.9-.7 4.9-2.6 4.6-4.6-.3-1.9-3-3.2-5.9-2.9M244.8 8C106.1 8 0 113.3 0 252c0 110.9 69.8 205.8 169.5 239.2 12.8 2.3 17.3-5.6 17.3-12.1 0-6.2-.3-40.4-.3-61.4 0 0-70 15-84.7-29.8 0 0-11.4-29.1-27.8-36.6 0 0-22.9-15.7 1.6-15.4 0 0 24.9 2 38.6 25.8 21.9 38.6 58.6 27.5 72.9 20.9 2.3-16 8.8-27.1 16-33.7-55.9-6.2-112.3-14.3-112.3-110.5 0-27.5 7.6-41.3 23.6-58.9-2.6-6.5-11.1-33.3 2.6-67.9 20.9-6.5 69 27 69 27 20-5.6 41.5-8.5 62.8-8.5s42.8 2.9 62.8 8.5c0 0 48.1-33.6 69-27 13.7 34.7 5.2 61.4 2.6 67.9 16 17.7 25.8 31.5 25.8 58.9 0 96.5-58.9 104.2-114.8 110.5 9.2 7.9 17 22.9 17 46.4 0 33.7-.3 75.4-.3 83.6 0 6.5 4.6 14.4 17.3 12.1C428.2 457.8 496 362.9 496 252 496 113.3 383.5 8 244.8 8M97.2 352.9c-1.3 1-1 3.3.7 5.2 1.6 1.6 3.9 2.3 5.2 1 1.3-1 1-3.3-.7-5.2-1.6-1.6-3.9-2.3-5.2-1m-10.8-8.1c-.7 1.3.3 2.9 2.3 3.9 1.6 1 3.6.7 4.3-.7.7-1.3-.3-2.9-2.3-3.9-2-.6-3.6-.3-4.3.7m32.4 35.6c-1.6 1.3-1 4.3 1.3 6.2 2.3 2.3 5.2 2.6 6.5 1 1.3-1.3.7-4.3-1.3-6.2-2.2-2.3-5.2-2.6-6.5-1m-11.4-14.7c-1.6 1-1.6 3.6 0 5.9s4.3 3.3 5.6 2.3c1.6-1.3 1.6-3.9 0-6.2-1.4-2.3-4-3.3-5.6-2"/></svg>
  </div>
  <div class="md-source__repository">
    genomic-medicine-sweden/nallo
  </div>
</a>
    </div>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href=".." class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Home
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../usage/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Usage
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
    
  
  
  
    <li class="md-nav__item md-nav__item--active">
      
      <input class="md-nav__toggle md-toggle" type="checkbox" id="__toc">
      
      
        
      
      
        <label class="md-nav__link md-nav__link--active" for="__toc">
          
  
  <span class="md-ellipsis">
    Output
  </span>
  

          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <a href="./" class="md-nav__link md-nav__link--active">
        
  
  <span class="md-ellipsis">
    Output
  </span>
  

      </a>
      
        

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#table-of-contents" class="md-nav__link">
    <span class="md-ellipsis">
      Table of contents
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#pipeline-overview" class="md-nav__link">
    <span class="md-ellipsis">
      Pipeline overview
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Pipeline overview">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#alignment" class="md-nav__link">
    <span class="md-ellipsis">
      Alignment
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#assembly" class="md-nav__link">
    <span class="md-ellipsis">
      Assembly
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#assembly-variant-calling" class="md-nav__link">
    <span class="md-ellipsis">
      Assembly variant calling
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#cnv-calling" class="md-nav__link">
    <span class="md-ellipsis">
      CNV calling
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#methylation" class="md-nav__link">
    <span class="md-ellipsis">
      Methylation
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#multiqc" class="md-nav__link">
    <span class="md-ellipsis">
      MultiQC
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#paraphase" class="md-nav__link">
    <span class="md-ellipsis">
      Paraphase
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#phasing" class="md-nav__link">
    <span class="md-ellipsis">
      Phasing
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#pipeline-information" class="md-nav__link">
    <span class="md-ellipsis">
      Pipeline information
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#qc" class="md-nav__link">
    <span class="md-ellipsis">
      QC
    </span>
  </a>
  
    <nav class="md-nav" aria-label="QC">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#fastqc" class="md-nav__link">
    <span class="md-ellipsis">
      FastQC
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#mosdepth" class="md-nav__link">
    <span class="md-ellipsis">
      Mosdepth
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#cramino" class="md-nav__link">
    <span class="md-ellipsis">
      Cramino
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#somalier" class="md-nav__link">
    <span class="md-ellipsis">
      Somalier
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
          <li class="md-nav__item">
  <a href="#repeat-calling" class="md-nav__link">
    <span class="md-ellipsis">
      Repeat calling
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#repeat-annotation" class="md-nav__link">
    <span class="md-ellipsis">
      Repeat annotation
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#snvs" class="md-nav__link">
    <span class="md-ellipsis">
      SNVs
    </span>
  </a>
  
    <nav class="md-nav" aria-label="SNVs">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#calling" class="md-nav__link">
    <span class="md-ellipsis">
      Calling
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#annotation" class="md-nav__link">
    <span class="md-ellipsis">
      Annotation
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ranking" class="md-nav__link">
    <span class="md-ellipsis">
      Ranking
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
          <li class="md-nav__item">
  <a href="#sv-calling" class="md-nav__link">
    <span class="md-ellipsis">
      SV Calling
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#sv-annotation" class="md-nav__link">
    <span class="md-ellipsis">
      SV Annotation
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
    </ul>
  
</nav>
      
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../parameters/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Parameters
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../CITATIONS/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Citations
  </span>
  

      </a>
    </li>
  

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
              
              <div class="md-sidebar md-sidebar--secondary" data-md-component="sidebar" data-md-type="toc" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#table-of-contents" class="md-nav__link">
    <span class="md-ellipsis">
      Table of contents
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#pipeline-overview" class="md-nav__link">
    <span class="md-ellipsis">
      Pipeline overview
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Pipeline overview">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#alignment" class="md-nav__link">
    <span class="md-ellipsis">
      Alignment
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#assembly" class="md-nav__link">
    <span class="md-ellipsis">
      Assembly
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#assembly-variant-calling" class="md-nav__link">
    <span class="md-ellipsis">
      Assembly variant calling
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#cnv-calling" class="md-nav__link">
    <span class="md-ellipsis">
      CNV calling
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#methylation" class="md-nav__link">
    <span class="md-ellipsis">
      Methylation
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#multiqc" class="md-nav__link">
    <span class="md-ellipsis">
      MultiQC
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#paraphase" class="md-nav__link">
    <span class="md-ellipsis">
      Paraphase
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#phasing" class="md-nav__link">
    <span class="md-ellipsis">
      Phasing
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#pipeline-information" class="md-nav__link">
    <span class="md-ellipsis">
      Pipeline information
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#qc" class="md-nav__link">
    <span class="md-ellipsis">
      QC
    </span>
  </a>
  
    <nav class="md-nav" aria-label="QC">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#fastqc" class="md-nav__link">
    <span class="md-ellipsis">
      FastQC
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#mosdepth" class="md-nav__link">
    <span class="md-ellipsis">
      Mosdepth
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#cramino" class="md-nav__link">
    <span class="md-ellipsis">
      Cramino
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#somalier" class="md-nav__link">
    <span class="md-ellipsis">
      Somalier
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
          <li class="md-nav__item">
  <a href="#repeat-calling" class="md-nav__link">
    <span class="md-ellipsis">
      Repeat calling
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#repeat-annotation" class="md-nav__link">
    <span class="md-ellipsis">
      Repeat annotation
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#snvs" class="md-nav__link">
    <span class="md-ellipsis">
      SNVs
    </span>
  </a>
  
    <nav class="md-nav" aria-label="SNVs">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#calling" class="md-nav__link">
    <span class="md-ellipsis">
      Calling
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#annotation" class="md-nav__link">
    <span class="md-ellipsis">
      Annotation
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ranking" class="md-nav__link">
    <span class="md-ellipsis">
      Ranking
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
          <li class="md-nav__item">
  <a href="#sv-calling" class="md-nav__link">
    <span class="md-ellipsis">
      SV Calling
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#sv-annotation" class="md-nav__link">
    <span class="md-ellipsis">
      SV Annotation
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
    </ul>
  
</nav>
                  </div>
                </div>
              </div>
            
          
          
            <div class="md-content" data-md-component="content">
              <article class="md-content__inner md-typeset">
                
                  

  
  


<h1 id="genomic-medicine-swedennallo-output">genomic-medicine-sweden/nallo: Output</h1>
<h2 id="table-of-contents">Table of contents</h2>
<ul>
<li><a href="#genomic-medicine-swedennallo-output">genomic-medicine-sweden/nallo: Output</a></li>
<li><a href="#table-of-contents">Table of contents</a></li>
<li><a href="#pipeline-overview">Pipeline overview</a><ul>
<li><a href="#alignment">Alignment</a></li>
<li><a href="#assembly">Assembly</a></li>
<li><a href="#assembly-variant-calling">Assembly variant calling</a></li>
<li><a href="#cnv-calling">CNV calling</a></li>
<li><a href="#methylation">Methylation</a></li>
<li><a href="#multiqc">MultiQC</a></li>
<li><a href="#paraphase">Paraphase</a></li>
<li><a href="#phasing">Phasing</a></li>
<li><a href="#pipeline-information">Pipeline information</a></li>
<li><a href="#qc">QC</a></li>
<li><a href="#fastqc">FastQC</a></li>
<li><a href="#mosdepth">Mosdepth</a></li>
<li><a href="#cramino">Cramino</a></li>
<li><a href="#somalier">Somalier</a></li>
<li><a href="#repeat-calling">Repeat calling</a></li>
<li><a href="#repeat-annotation">Repeat annotation</a></li>
<li><a href="#snvs">SNVs</a></li>
<li><a href="#calling">Calling</a></li>
<li><a href="#annotation">Annotation</a></li>
<li><a href="#ranking">Ranking</a></li>
<li><a href="#ranked-variants">Ranked Variants</a></li>
<li><a href="#sv-calling">SV Calling</a></li>
<li><a href="#sv-annotation">SV Annotation</a></li>
</ul>
</li>
</ul>
<h2 id="pipeline-overview">Pipeline overview</h2>
<p>The directories listed below will be created in the results directory after the pipeline has finished:</p>
<ul>
<li><code>aligned_reads</code></li>
<li><code>assembly_haplotypes</code></li>
<li><code>assembly_variant_calling</code></li>
<li><code>cnv_calling</code></li>
<li><code>databases</code></li>
<li><code>methylation</code></li>
<li><code>multiqc</code></li>
<li><code>paraphase</code></li>
<li><code>pedigree</code></li>
<li><code>phasing</code></li>
<li><code>pipeline_info</code></li>
<li><code>qc</code></li>
<li><code>repeat_calling</code></li>
<li><code>snvs</code></li>
<li><code>svs</code></li>
</ul>
<h3 id="alignment">Alignment</h3>
<p><a href="https://github.com/lh3/minimap2">minimap2</a> is used to map the reads to a reference genome. The aligned reads are sorted, (merged) and indexed using <a href="https://github.com/samtools/samtools">samtools</a>.</p>
<details>
<summary>Output files from Alignment</summary>
<ul>
<li><code>{outputdir}/aligned_reads/minimap2/{sample}/</code></li>
<li><code>*.bam</code>: Alignment file in bam format</li>
<li><code>*.bai</code>: Index of the corresponding bam file
  </li>
</ul>
</details>
<h3 id="assembly">Assembly</h3>
<p><a href="https://github.com/chhylp123/hifiasm">hifiasm</a> is used to assemble genomes. The assembled haplotypes are then comverted to fasta files using <a href="https://github.com/vgl-hub/gfastats">gfastats</a>.</p>
<details>
<summary>Output files from Assembly</summary>
<ul>
<li><code>{outputdir}/assembly_haplotypes/gfastats/{sample}/</code></li>
<li><code>*hap1.p_ctg.fasta.gz</code>: Assembled haplotype 1</li>
<li><code>*hap2.p_ctg.fasta.gz</code>: Assembled haplotype 2</li>
<li><code>*.assembly_summary</code>: Summary statistics
  </li>
</ul>
</details>
<h3 id="assembly-variant-calling">Assembly variant calling</h3>
<p>A deconstructed version of <a href="https://github.com/lh3/dipcall">dipcall</a> is used to call variants from the assembled haplotypes. They are also mapped back to the reference genome.</p>
<details>
<summary>Output files from Assembly variant calling</summary>
<blockquote>
<p>Dipcall produces several files, a full expanation is available <a href="https://github.com/lh3/dipcall">here</a>.</p>
</blockquote>
<ul>
<li>
<p><code>{outputdir}/assembly_variant_calling/dipcall/{sample}/</code></p>
</li>
<li>
<p><code>*hap1.bam</code>: Assembled haplotype 1 mapped to the reference genome</p>
</li>
<li><code>*hap1.bai</code>: Index of the corresponding bam file.</li>
<li><code>*hap2.bam</code>: Assembled haplotype 2 mapped to the reference genome</li>
<li><code>*hap2.bai</code>: Index of the corresponding bam file.</li>
</ul>
</details>
<h3 id="cnv-calling">CNV calling</h3>
<p><a href="https://github.com/PacificBiosciences/HiFiCNV">HiFiCNV</a> is used to call CNVs. It also produces copynumber, depth and MAF tracks loadable in IGV.</p>
<details>
<summary>Output files from CNV calling</summary>
<ul>
<li><code>{outputdir}/cnv_calling/hificnv/{sample}/</code></li>
<li><code>*.copynum.bedgraph</code>: Copy number in bedgraph format</li>
<li><code>*.depth.bw</code>: Depth track in BigWig format</li>
<li><code>*.maf.bw</code>: Minor allele frequencies in BigWig format</li>
<li><code>*.vcf.gz</code>: VCF file containing CNV variants</li>
<li><code>*.vcf.gz.tbi</code>: Index of the corresponding VCF file
  </li>
</ul>
</details>
<h3 id="methylation">Methylation</h3>
<p><a href="https://github.com/nanoporetech/modkit">modkit</a> is used to create methylation pileups. bedMethyl files are stored both one file with summary counts from reads per haplotag (e.g. HP1, HP2 and ungrouped) and one file with summary counts from all reads. The methylation is also stored in the BAM files and can be viewed directly in IGV.</p>
<details>
<summary>Output files from Methylation</summary>
<ul>
<li>
<p><code>{outputdir}/methylation/modkit/pileup/phased/{sample}/</code></p>
</li>
<li>
<p><code>*.modkit_pileup_phased_*.bed.gz</code>: bedMethyl file containing summary counts from reads with haplotags, e.g. 1 or 2</p>
</li>
<li><code>*.modkit_pileup_phased_ungrouped.bed.gz</code>: bedMethyl file containing summary counts for ungrouped reads</li>
<li>
<p><code>*.bed.gz.tbi</code>: Index of the corresponding bedMethyl file</p>
</li>
<li>
<p><code>{outputdir}/methylation/modkit/pileup/unphased/{sample}/</code></p>
</li>
<li><code>*.modkit_pileup.bed.gz</code>: bedMethyl file containing summary counts from all reads</li>
<li><code>*.bed.gz.tbi</code>: Index of the corresponding bedMethyl file
  </li>
</ul>
</details>
<h3 id="multiqc">MultiQC</h3>
<p><a href="http://multiqc.info">MultiQC</a> is a visualization tool that generates a single HTML report summarising all samples in your project. Most of the pipeline QC results are visualised in the report and further statistics are available in the report data directory.</p>
<p>Results generated by MultiQC collate pipeline QC from supported tools e.g. FastQC. The pipeline has special steps which also allow the software versions to be reported in the MultiQC output for future traceability. For more information about how to use MultiQC reports, see <a href="http://multiqc.info">http://multiqc.info</a>.</p>
<details>
<summary>Output files</summary>
<ul>
<li><code>{outputdir}/multiqc/</code></li>
<li><code>multiqc_report.html</code>: a standalone HTML file that can be viewed in your web browser.</li>
<li><code>multiqc_data/</code>: directory containing parsed statistics from the different tools used in the pipeline.</li>
<li><code>multiqc_plots/</code>: directory containing static images from the report in various formats.
  </li>
</ul>
</details>
<h3 id="paraphase">Paraphase</h3>
<p><a href="https://github.com/PacificBiosciences/paraphase">Paraphase</a> is used to call paralogous genes. For interpreting the output, see <a href="https://github.com/PacificBiosciences/paraphase">https://github.com/PacificBiosciences/paraphase</a>.</p>
<details>
<summary>Output files from Paraphase</summary>
<ul>
<li><code>{outputdir}/paraphase/{sample}/</code></li>
<li><code>*.bam</code>: BAM file with haplotypes grouped by HP and colored by YC</li>
<li><code>*.bai</code>: Index of the corresponding bam file.</li>
<li><code>*.json</code>: Output file summarizing haplotypes and variant calls</li>
<li><code>{sample}_paraphase_vcfs/</code>:<ul>
<li><code>{sample}_{gene}_vcf</code>: VCF file per gene</li>
<li><code>{sample}_{gene}_vcf.tbi</code>: Index of the corresponding VCF file
</li>
</ul>
</li>
</ul>
</details>
<h3 id="phasing">Phasing</h3>
<p><a href="https://github.com/twolinin/longphase">LongPhase</a>, <a href="https://whatshap.readthedocs.io/en/latest/">WhatsHap</a> or <a href="https://github.com/PacificBiosciences/HiPhase">HiPhase</a> are used to phase variants and haplotag reads.</p>
<details>
<summary>Output files from phasing</summary>
<ul>
<li><code>{outputdir}/aligned_reads/{sample}/</code></li>
<li><code>{sample}_haplotagged.bam</code>: BAM file with haplotags</li>
<li><code>{sample}_haplotagged.bam.bai</code>: Index of the corresponding bam file</li>
<li><code>{outputdir}/phased_variants/{sample}/</code></li>
<li><code>*.vcf.gz</code>: VCF file with phased variants</li>
<li><code>*.vcf.gz.tbi</code>: Index of the corresponding VCF file</li>
<li><code>{outputdir}/qc/phasing_stats/{sample}/</code></li>
<li><code>*.blocks.tsv</code>: File with phase blocks</li>
<li><code>*.stats.tsv</code>: File with phasing statistics
  </li>
</ul>
</details>
<h3 id="pipeline-information">Pipeline information</h3>
<p><a href="https://www.nextflow.io/docs/latest/tracing.html">Nextflow</a> provides excellent functionality for generating various reports relevant to the running and execution of the pipeline. This will allow you to troubleshoot errors with the running of the pipeline, and also provide you with other information such as launch commands, run times and resource usage.</p>
<details>
<summary>Output files</summary>
<ul>
<li><code>pipeline_info/</code></li>
<li>Reports generated by Nextflow: <code>execution_report.html</code>, <code>execution_timeline.html</code>, <code>execution_trace.txt</code> and <code>pipeline_dag.dot</code>/<code>pipeline_dag.svg</code>.</li>
<li>Reports generated by the pipeline: <code>pipeline_report.html</code>, <code>pipeline_report.txt</code> and <code>software_versions.yml</code>. The <code>pipeline_report*</code> files will only be present if the <code>--email</code> / <code>--email_on_fail</code> parameter's are used when running the pipeline.</li>
<li>Reformatted samplesheet files used as input to the pipeline: <code>samplesheet.valid.csv</code>.</li>
<li>Parameters used by the pipeline run: <code>params.json</code>.</li>
</ul>
</details>
<h3 id="qc">QC</h3>
<p><a href="http://www.bioinformatics.babraham.ac.uk/projects/fastqc/">FastQC</a>, <a href="https://github.com/wdecoster/cramino">cramino</a>, <a href="https://github.com/brentp/mosdepth">mosdepth</a> and <a href="https://github.com/brentp/somalier">somalier</a> are used for read QC.</p>
<h5 id="fastqc">FastQC</h5>
<p><a href="http://www.bioinformatics.babraham.ac.uk/projects/fastqc/">FastQC</a> gives general quality metrics about your sequenced reads. It provides information about the quality score distribution across your reads, per base sequence content (%A/T/G/C), adapter contamination and overrepresented sequences. For further reading and documentation see the <a href="http://www.bioinformatics.babraham.ac.uk/projects/fastqc/Help/">FastQC help pages</a>.</p>
<details>
<summary>Output files</summary>
<ul>
<li><code>{outputdir}/qc/fastqc/{sample}/</code></li>
<li><code>*_fastqc.html</code>: FastQC report containing quality metrics.</li>
<li><code>*_fastqc.zip</code>: Zip archive containing the FastQC report, tab-delimited data file and plot images.
  </li>
</ul>
</details>
<h5 id="mosdepth">Mosdepth</h5>
<p><a href="https://github.com/brentp/mosdepth">Mosdepth</a> is used to report quality control metrics such as coverage, and GC content from alignment files.</p>
<details>
<summary>Output files from Mosdepth</summary>
<ul>
<li><code>{outputdir}/qc/mosdepth/{sample}</code></li>
<li><code>*.mosdepth.global.dist.txt</code>: This file contains a cumulative distribution indicating the proportion of total bases that were covered for at least a given coverage value across each chromosome and the whole genome</li>
<li><code>*.mosdepth.region.dist.txt</code>: This file contains a cumulative distribution indicating the proportion of total bases that were covered for at least a given coverage value across each region, is output if running the pipeline with a BED-file</li>
<li><code>*.mosdepth.summary.txt</code>: Mosdepth ummary file</li>
<li><code>*.regions.bed.gz</code>: Depth per region, is output if running the pipeline with a BED-file</li>
<li><code>*.regions.bed.gz.csi</code>: Index of regions.bed.gz
  </li>
</ul>
</details>
<h5 id="cramino">Cramino</h5>
<p><a href="https://github.com/wdecoster/cramino">cramino</a> is run on both phased and unphased reads.</p>
<details>
<summary>Output files from Cramino</summary>
<ul>
<li><code>{outputdir}/qc/cramino/phased/{sample}</code></li>
<li><code>*.arrow</code>: Read length and quality in <a href="https://arrow.apache.org/docs/format/Columnar.html">Apache Arrow</a> format</li>
<li><code>*.txt</code>: Summary information in text format</li>
<li><code>{outputdir}/qc/cramino/unphased/{sample}</code></li>
<li><code>*.arrow</code>: Read length and quality in <a href="https://arrow.apache.org/docs/format/Columnar.html">Apache Arrow</a> format</li>
<li><code>*.txt</code>: Summary information in text format
  </li>
</ul>
</details>
<h5 id="somalier">Somalier</h5>
<p><a href="https://github.com/brentp/somalier">somalier</a> is used to check relatedness and sex.</p>
<details>
<summary>Output files from Somalier</summary>
<ul>
<li><code>{outputdir}/predigree/{project}.ped</code>: A PED file with updated from somalier sex</li>
<li><code>{outputdir}/qc/somalier/relate/{project}/</code></li>
<li><code>{project}.html</code>: HTML report</li>
<li><code>{project}.pairs.tsv</code>: Output information in sample pairs</li>
<li><code>{project}.samples.tsv</code>: Output information per sample
  </li>
</ul>
</details>
<h3 id="repeat-calling">Repeat calling</h3>
<p><a href="https://github.com/PacificBiosciences/trgt">TRGT</a> is used to call repeats.</p>
<details>
<summary>Output files from TRGT</summary>
<ul>
<li><code>{outputdir}/repeat_calling/trgt/multi_sample/{project}/</code></li>
<li><code>*.vcf.gz</code>: Merged VCF for all samples</li>
<li><code>*.vcf.gz.tbi</code>: Index of the corresponding VCF file</li>
<li><code>{outputdir}/repeat_calling/trgt/single_sample/{sample}/</code></li>
<li><code>*.vcf.gz</code>: VCF with called repeats</li>
<li><code>*.vcf.gz.tbi</code>: Index of the corresponding VCF file</li>
<li><code>*.bam</code>: BAM file with sorted spanning reads</li>
<li><code>*.bai</code>: Index of the corresponding bam file
  </li>
</ul>
</details>
<h3 id="repeat-annotation">Repeat annotation</h3>
<p><a href="https://github.com/Clinical-Genomics/stranger">Stranger</a> is used to annotate repeats.</p>
<details>
<summary>Output files from Stranger</summary>
<ul>
<li><code>{outputdir}/repeat_annotation/stranger/{sample}</code></li>
<li><code>*.vcf.gz</code>: Annotated VCF</li>
<li><code>*.vcf.gz.tbi</code>: Index of the corresponding VCF file
  </li>
</ul>
</details>
<h3 id="snvs">SNVs</h3>
<h4 id="calling">Calling</h4>
<p><a href="https://github.com/google/deepvariant">DeepVariant</a> is used to call variants, <a href="https://samtools.github.io/bcftools/bcftools.html">bcftools</a> and <a href="https://github.com/dnanexus-rnd/GLnexus">GLnexus</a> are used to merge variants.</p>
<details>
<summary>Output files from SNV calling</summary>
<blockquote>
<p>[!NOTE]
Variants are only output without annotation and ranking if these subworkflows are turned off.</p>
</blockquote>
<ul>
<li><code>{outputdir}/snvs/single_sample/{sample}/</code></li>
<li><code>{sample}_snv.vcf.gz</code>: VCF with called variants with alternative genotypes from a certain sample</li>
<li><code>{sample}_snv.vcf.gz.tbi</code>: Index of the corresponding VCF file</li>
<li><code>{outputdir}/snvs/multi_sample/{project}/</code></li>
<li><code>{project}_snv.vcf.gz</code>: VCF with called variants from all samples</li>
<li><code>{project}_snv.vcf.gz.tbi</code>: Index of the corresponding VCF file</li>
<li><code>{outputdir}/snvs/stats/single_sample/</code></li>
<li><code>*.stats.txt</code>: Variant statistics
  </li>
</ul>
</details>
<h4 id="annotation">Annotation</h4>
<p><a href="https://github.com/brentp/echtvar">echtvar</a> and <a href="https://www.ensembl.org/vep">VEP</a> are used to annotate SNVs. <a href="https://cadd.gs.washington.edu/">CADD</a> is used to annotate INDELs with CADD scores.</p>
<details>
<summary>Output files from SNV Annotation</summary>
<blockquote>
<p>[!NOTE]
Variants are only output without ranking if that subworkflows are turned off.</p>
</blockquote>
<ul>
<li><code>{outputdir}/databases/echtvar/encode/{project}/</code></li>
<li><code>*.zip</code>: Database with AF and AC for all samples run</li>
<li><code>{outputdir}/snvs/single_sample/{sample}/</code></li>
<li><code>{sample}_snv_annotated.vcf.gz</code>: VCF with annotated variants with alternative genotypes from a certain sample</li>
<li><code>{sample}_snv_annotated.vcf.gz.tbi</code>: Index of the corresponding VCF file</li>
<li><code>{outputdir}/snvs/multi_sample/{project}/</code></li>
<li><code>{project}_snv_annotated.vcf.gz</code>: VCF with annotated variants from all samples</li>
<li><code>{project}_snv_annotated.vcf.gz.tbi</code>: Index of the corresponding VCF file
  </li>
</ul>
</details>
<h4 id="ranking">Ranking</h4>
<p><a href="https://github.com/Clinical-Genomics/genmod">GENMOD</a> is a simple to use command line tool for annotating and analyzing genomic variations in the VCF file format. GENMOD can annotate genetic patterns of inheritance in vcf files with single or multiple families of arbitrary size. Each variant will be assigned a predicted pathogenicity score. The score will be given both as a raw score and a normalized score with values between 0 and 1. The tags in the INFO field are <code>RankScore</code> and <code>RankScoreNormalized</code>. The score can be configured to fit your annotations and preferences by modifying the score config file.</p>
<details>
<summary>Output files from SNV ranking</summary>
<ul>
<li><code>{outputdir}/snvs/single_sample/{sample}/</code></li>
<li><code>{sample}_snv_annotated_ranked.vcf.gz</code>: VCF with annotated and ranked variants with alternative genotypes from a certain sample</li>
<li><code>{sample}_snv_annotated_ranked.vcf.gz.tbi</code>: Index of the corresponding VCF file</li>
<li><code>{outputdir}/snvs/multi_sample/{project}/</code></li>
<li><code>{project}_snv_annotated_ranked.vcf.gz</code>: VCF with annotated and ranked variants from all samples</li>
<li><code>{project}_snv_annotated_ranked.vcf.gz.tbi</code>: Index of the corresponding VCF file
  </li>
</ul>
</details>
<h3 id="sv-calling">SV Calling</h3>
<p><a href="https://github.com/KolmogorovLab/Severus">Severus</a> or <a href="https://github.com/fritzsedlazeck/Sniffles">Sniffles</a> is used to call structural variants, after wich <a href="https://github.com/J35P312/SVDB">SVDB</a> is used to merge variants, within and between samples.</p>
<details>
<summary>Output files from SV Calling</summary>
<ul>
<li><code>{outputdir}/svs/multi_sample/{project}</code></li>
<li><code>{project}_svs.vcf.gz</code>: VCF file with SVDB merged variants</li>
<li><code>{project}_svs.vcf.gz.tbi</code>: Index of the corresponding VCF file</li>
<li><code>{outputdir}/svs/single_sample/{sample}</code></li>
<li><code>*.vcf.gz</code>: VCF with SVDB merged variants, divided per sample</li>
<li><code>*.vcf.gz.tbi</code>: Index of the corresponding VCF file
  </li>
</ul>
</details>
<h3 id="sv-annotation">SV Annotation</h3>
<p><a href="https://github.com/J35P312/SVDB">SVDB</a> and <a href="https://www.ensembl.org/vep">VEP</a> are used to annotate SVs.</p>
<details>
<summary>Output files from SV Annotation</summary>
<ul>
<li><code>{outputdir}/svs/multi_sample/{project}</code></li>
<li><code>{project}_svs_annotated.vcf.gz</code>: VCF file with annotated merged variants</li>
<li><code>{project}_svs_annotated.vcf.gz.tbi</code>: Index of the corresponding VCF file</li>
<li><code>{outputdir}/svs/single_sample/{sample}</code></li>
<li><code>*.vcf_annotated.gz</code>: VCF with annotated variants per sample</li>
<li><code>*.vcf_annotated.gz.tbi</code>: Index of the corresponding VCF file
  </li>
</ul>
</details>












                
              </article>
            </div>
          
          
<script>var target=document.getElementById(location.hash.slice(1));target&&target.name&&(target.checked=target.name.startsWith("__tabbed_"))</script>
        </div>
        
          <button type="button" class="md-top md-icon" data-md-component="top" hidden>
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M13 20h-2V8l-5.5 5.5-1.42-1.42L12 4.16l7.92 7.92-1.42 1.42L13 8z"/></svg>
  Back to top
</button>
        
      </main>
      
        <footer class="md-footer">
  
    
      
      <nav class="md-footer__inner md-grid" aria-label="Footer" >
        
          
          <a href="../usage/" class="md-footer__link md-footer__link--prev" aria-label="Previous: Usage">
            <div class="md-footer__button md-icon">
              
              <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11z"/></svg>
            </div>
            <div class="md-footer__title">
              <span class="md-footer__direction">
                Previous
              </span>
              <div class="md-ellipsis">
                Usage
              </div>
            </div>
          </a>
        
        
          
          <a href="../parameters/" class="md-footer__link md-footer__link--next" aria-label="Next: Parameters">
            <div class="md-footer__title">
              <span class="md-footer__direction">
                Next
              </span>
              <div class="md-ellipsis">
                Parameters
              </div>
            </div>
            <div class="md-footer__button md-icon">
              
              <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M4 11v2h12l-5.5 5.5 1.42 1.42L19.84 12l-7.92-7.92L10.5 5.5 16 11z"/></svg>
            </div>
          </a>
        
      </nav>
    
  
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-copyright">
  
  
    Made with
    <a href="https://squidfunk.github.io/mkdocs-material/" target="_blank" rel="noopener">
      Material for MkDocs
    </a>
  
</div>
      
    </div>
  </div>
</footer>
      
    </div>
    <div class="md-dialog" data-md-component="dialog">
      <div class="md-dialog__inner md-typeset"></div>
    </div>
    
    
    <script id="__config" type="application/json">{"base": "..", "features": ["content.code.annotate", "content.code.copy", "navigation.instant", "navigation.top", "navigation.tracking", "navigation.sections", "navigation.footer", "search.share", "toc.follow"], "search": "../assets/javascripts/workers/search.6ce7567c.min.js", "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version": "Select version"}, "version": {"provider": "mike"}}</script>
    
    
      <script src="../assets/javascripts/bundle.83f73b43.min.js"></script>
      
    
  </body>
</html>