Changes in module 'nf-core/bcftools/view'
--- modules/nf-core/bcftools/view/main.nf
+++ modules/nf-core/bcftools/view/main.nf
@@ -8,10 +8,7 @@
         'biocontainers/bcftools:1.17--haef29d1_0' }"
 
     input:
-    tuple val(meta), path(vcf), path(index)
-    path(regions)
-    path(targets)
-    path(samples)
+    tuple val(meta), path(vcf)
 
     output:
     tuple val(meta), path("*.gz") , emit: vcf
@@ -23,14 +20,11 @@
     script:
     def args = task.ext.args ?: ''
     def prefix = task.ext.prefix ?: "${meta.id}"
-    def regions_file  = regions ? "--regions-file ${regions}" : ""
-    def targets_file = targets ? "--targets-file ${targets}" : ""
-    def samples_file =  samples ? "--samples-file ${samples}" : ""
+    // Exlude samples in samples 
+    def samples_file = "--samples ^${meta.id}"
     """
     bcftools view \\
         --output ${prefix}.vcf.gz \\
-        ${regions_file} \\
-        ${targets_file} \\
         ${samples_file} \\
         $args \\
         --threads $task.cpus \\

************************************************************
