Changes in module 'nf-core/samtools/faidx'
--- modules/nf-core/samtools/faidx/main.nf
+++ modules/nf-core/samtools/faidx/main.nf
@@ -2,10 +2,10 @@
     tag "$fasta"
     label 'process_single'
 
-    conda "bioconda::samtools=1.16.1"
+    conda "bioconda::samtools=1.17"
     container "${ workflow.containerEngine == 'singularity' && !task.ext.singularity_pull_docker_container ?
-        'https://depot.galaxyproject.org/singularity/samtools:1.16.1--h6899075_1' :
-        'quay.io/biocontainers/samtools:1.16.1--h6899075_1' }"
+        'https://depot.galaxyproject.org/singularity/samtools:1.17--h00cdaf9_0'' :
+        'quay.io/biocontainers/samtools:1.17--h00cdaf9_0'' }"
 
     input:
     tuple val(meta), path(fasta)

************************************************************
