Changes in module 'nf-core/cadd'
--- modules/nf-core/cadd/main.nf
+++ modules/nf-core/cadd/main.nf
@@ -43,7 +43,7 @@
     def prefix = task.ext.prefix ?: "${meta.id}"
     def VERSION = "1.6.post1" // WARN: Version information not provided by tool on CLI. Please update version string below when bumping container versions.
     """
-    touch ${prefix}.tsv.gz
+    echo "" | gzip > ${prefix}.tsv.gz
 
     cat <<-END_VERSIONS > versions.yml
     "${task.process}":

************************************************************
