Changes in module 'nf-core/tabix/bgziptabix'
--- modules/nf-core/tabix/bgziptabix/main.nf
+++ modules/nf-core/tabix/bgziptabix/main.nf
@@ -23,8 +23,8 @@
     def args2 = task.ext.args2 ?: ''
     def prefix = task.ext.prefix ?: "${meta.id}"
     """
-    bgzip  --threads ${task.cpus} -c $args $input > ${prefix}.${input.getExtension()}.gz
-    tabix $args2 ${prefix}.${input.getExtension()}.gz
+    bgzip  --threads ${task.cpus} -c $args $input > ${input.baseName}.${input.getExtension()}.gz
+    tabix $args2 ${input.baseName}.${input.getExtension()}.gz
 
     cat <<-END_VERSIONS > versions.yml
     "${task.process}":

************************************************************
