Changes in module 'nf-core/gawk'
--- modules/nf-core/gawk/main.nf
+++ modules/nf-core/gawk/main.nf
@@ -12,7 +12,7 @@
     path(program_file)
 
     output:
-    tuple val(meta), path("${prefix}.${suffix}"), emit: output
+    tuple val(meta), path("*.txt"), emit: output
     path "versions.yml"                         , emit: versions
 
     when:
@@ -22,7 +22,7 @@
     def args  = task.ext.args  ?: '' // args is used for the main arguments of the tool
     def args2 = task.ext.args2 ?: '' // args2 is used to specify a program when no program file has been given
     prefix = task.ext.prefix ?: "${meta.id}"
-    suffix = task.ext.suffix ?: "${input.getExtension}"
+    suffix = task.ext.suffix ?: "awk.${input.Extension}"
 
     program = program_file ? "-f ${program_file}" : "${args2}"
 
@@ -30,8 +30,7 @@
     awk \\
         ${args} \\
         ${program} \\
-        ${input} \\
-        > ${prefix}.${suffix}
+        ${input}
 
     cat <<-END_VERSIONS > versions.yml
     "${task.process}":

************************************************************
