Changes in module 'nf-core/bcftools/merge'
--- modules/nf-core/bcftools/merge/main.nf
+++ modules/nf-core/bcftools/merge/main.nf
@@ -4,8 +4,8 @@
 
     conda "${moduleDir}/environment.yml"
     container "${ workflow.containerEngine == 'singularity' && !task.ext.singularity_pull_docker_container ?
-        'https://depot.galaxyproject.org/singularity/bcftools:1.18--h8b25389_0':
-        'biocontainers/bcftools:1.18--h8b25389_0' }"
+        'https://depot.galaxyproject.org/singularity/bcftools:1.20--h8b25389_0':
+        'biocontainers/bcftools:1.20--h8b25389_0' }"
 
     input:
     tuple val(meta), path(vcfs), path(tbis)

--- modules/nf-core/bcftools/merge/environment.yml
+++ modules/nf-core/bcftools/merge/environment.yml
@@ -4,4 +4,4 @@
   - bioconda
   - defaults
 dependencies:
-  - bioconda::bcftools=1.18
+  - bioconda::bcftools=1.20

************************************************************
