
<!doctype html>
<html lang="en" class="no-js">
  <head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      
      
      
        <link rel="canonical" href="https://genomic-medicine-sweden.github.io/nallo/0.6.0/output/">
      
      
        <link rel="prev" href="../usage/">
      
      
        <link rel="next" href="../parameters/">
      
      
      <link rel="icon" href="../assets/images/favicon.png">
      <meta name="generator" content="mkdocs-1.6.1, mkdocs-material-9.6.11">
    
    
      
        <title>Output - genomic-medicine-sweden/nallo</title>
      
    
    
      <link rel="stylesheet" href="../assets/stylesheets/main.4af4bdda.min.css">
      
        
        <link rel="stylesheet" href="../assets/stylesheets/palette.06af60db.min.css">
      
      


    
    
      
    
    
      
        
        
        <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Inter:300,300i,400,400i,700,700i%7CRoboto+Mono:400,400i,700,700i&display=fallback">
        <style>:root{--md-text-font:"Inter";--md-code-font:"Roboto Mono"}</style>
      
    
    
    <script>__md_scope=new URL("..",location),__md_hash=e=>[...e].reduce(((e,_)=>(e<<5)-e+_.charCodeAt(0)),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script>
    
      

    
    
    
  </head>
  
  
    
    
      
    
    
    
    
    <body dir="ltr" data-md-color-scheme="slate" data-md-color-primary="custom" data-md-color-accent="custom">
  
    
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off">
    <label class="md-overlay" for="__drawer"></label>
    <div data-md-component="skip">
      
        
        <a href="#genomic-medicine-swedennallo-output" class="md-skip">
          Skip to content
        </a>
      
    </div>
    <div data-md-component="announce">
      
    </div>
    
      <div data-md-color-scheme="default" data-md-component="outdated" hidden>
        
      </div>
    
    
      

  

<header class="md-header md-header--shadow" data-md-component="header">
  <nav class="md-header__inner md-grid" aria-label="Header">
    <a href=".." title="genomic-medicine-sweden/nallo" class="md-header__button md-logo" aria-label="genomic-medicine-sweden/nallo" data-md-component="logo">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 16 16"><path d="M2.5 1.75v11.5c0 .138.112.25.25.25h3.17a.75.75 0 0 1 0 1.5H2.75A1.75 1.75 0 0 1 1 13.25V1.75C1 .784 1.784 0 2.75 0h8.5C12.216 0 13 .784 13 1.75v7.736a.75.75 0 0 1-1.5 0V1.75a.25.25 0 0 0-.25-.25h-8.5a.25.25 0 0 0-.25.25m13.274 9.537zl-4.557 4.45a.75.75 0 0 1-1.055-.008l-1.943-1.95a.75.75 0 0 1 1.062-1.058l1.419 1.425 4.026-3.932a.75.75 0 1 1 1.048 1.074M4.75 4h4.5a.75.75 0 0 1 0 1.5h-4.5a.75.75 0 0 1 0-1.5M4 7.75A.75.75 0 0 1 4.75 7h2a.75.75 0 0 1 0 1.5h-2A.75.75 0 0 1 4 7.75"/></svg>

    </a>
    <label class="md-header__button md-icon" for="__drawer">
      
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M3 6h18v2H3zm0 5h18v2H3zm0 5h18v2H3z"/></svg>
    </label>
    <div class="md-header__title" data-md-component="header-title">
      <div class="md-header__ellipsis">
        <div class="md-header__topic">
          <span class="md-ellipsis">
            genomic-medicine-sweden/nallo
          </span>
        </div>
        <div class="md-header__topic" data-md-component="header-topic">
          <span class="md-ellipsis">
            
              Output
            
          </span>
        </div>
      </div>
    </div>
    
      
        <form class="md-header__option" data-md-component="palette">
  
    
    
    
    <input class="md-option" data-md-color-media="(prefers-color-scheme: dark)" data-md-color-scheme="slate" data-md-color-primary="custom" data-md-color-accent="custom"  aria-label="Switch to light mode"  type="radio" name="__palette" id="__palette_0">
    
      <label class="md-header__button md-icon" title="Switch to light mode" for="__palette_1" hidden>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 7a5 5 0 0 1 5 5 5 5 0 0 1-5 5 5 5 0 0 1-5-5 5 5 0 0 1 5-5m0 2a3 3 0 0 0-3 3 3 3 0 0 0 3 3 3 3 0 0 0 3-3 3 3 0 0 0-3-3m0-7 2.39 3.42C13.65 5.15 12.84 5 12 5s-1.65.15-2.39.42zM3.34 7l4.16-.35A7.2 7.2 0 0 0 5.94 8.5c-.44.74-.69 1.5-.83 2.29zm.02 10 1.76-3.77a7.131 7.131 0 0 0 2.38 4.14zM20.65 7l-1.77 3.79a7.02 7.02 0 0 0-2.38-4.15zm-.01 10-4.14.36c.59-.51 1.12-1.14 1.54-1.86.42-.73.69-1.5.83-2.29zM12 22l-2.41-3.44c.74.27 1.55.44 2.41.44.82 0 1.63-.17 2.37-.44z"/></svg>
      </label>
    
  
    
    
    
    <input class="md-option" data-md-color-media="(prefers-color-scheme: light)" data-md-color-scheme="default" data-md-color-primary="custom" data-md-color-accent="custom"  aria-label="Switch to dark mode"  type="radio" name="__palette" id="__palette_1">
    
      <label class="md-header__button md-icon" title="Switch to dark mode" for="__palette_0" hidden>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="m17.75 4.09-2.53 1.94.91 3.06-2.63-1.81-2.63 1.81.91-3.06-2.53-1.94L12.44 4l1.06-3 1.06 3zm3.5 6.91-1.64 1.25.59 1.98-1.7-1.17-1.7 1.17.59-1.98L15.75 11l2.06-.05L18.5 9l.69 1.95zm-2.28 4.95c.83-.08 1.72 1.1 1.19 1.85-.32.45-.66.87-1.08 1.27C15.17 23 8.84 23 4.94 19.07c-3.91-3.9-3.91-10.24 0-14.14.4-.4.82-.76 1.27-1.08.75-.53 1.93.36 1.85 1.19-.27 2.86.69 5.83 2.89 8.02a9.96 9.96 0 0 0 8.02 2.89m-1.64 2.02a12.08 12.08 0 0 1-7.8-3.47c-2.17-2.19-3.33-5-3.49-7.82-2.81 3.14-2.7 7.96.31 10.98 3.02 3.01 7.84 3.12 10.98.31"/></svg>
      </label>
    
  
</form>
      
    
    
      <script>var palette=__md_get("__palette");if(palette&&palette.color){if("(prefers-color-scheme)"===palette.color.media){var media=matchMedia("(prefers-color-scheme: light)"),input=document.querySelector(media.matches?"[data-md-color-media='(prefers-color-scheme: light)']":"[data-md-color-media='(prefers-color-scheme: dark)']");palette.color.media=input.getAttribute("data-md-color-media"),palette.color.scheme=input.getAttribute("data-md-color-scheme"),palette.color.primary=input.getAttribute("data-md-color-primary"),palette.color.accent=input.getAttribute("data-md-color-accent")}for(var[key,value]of Object.entries(palette.color))document.body.setAttribute("data-md-color-"+key,value)}</script>
    
    
    
      
      
        <label class="md-header__button md-icon" for="__search">
          
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"/></svg>
        </label>
        <div class="md-search" data-md-component="search" role="dialog">
  <label class="md-search__overlay" for="__search"></label>
  <div class="md-search__inner" role="search">
    <form class="md-search__form" name="search">
      <input type="text" class="md-search__input" name="query" aria-label="Search" placeholder="Search" autocapitalize="off" autocorrect="off" autocomplete="off" spellcheck="false" data-md-component="search-query" required>
      <label class="md-search__icon md-icon" for="__search">
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"/></svg>
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11z"/></svg>
      </label>
      <nav class="md-search__options" aria-label="Search">
        
          <a href="javascript:void(0)" class="md-search__icon md-icon" title="Share" aria-label="Share" data-clipboard data-clipboard-text="" data-md-component="search-share" tabindex="-1">
            
            <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M18 16.08c-.76 0-1.44.3-1.96.77L8.91 12.7c.05-.23.09-.46.09-.7s-.04-.47-.09-.7l7.05-4.11c.54.5 1.25.81 2.04.81a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3c0 .24.04.47.09.7L8.04 9.81C7.5 9.31 6.79 9 6 9a3 3 0 0 0-3 3 3 3 0 0 0 3 3c.79 0 1.5-.31 2.04-.81l7.12 4.15c-.05.21-.08.43-.08.66 0 1.61 1.31 2.91 2.92 2.91s2.92-1.3 2.92-2.91A2.92 2.92 0 0 0 18 16.08"/></svg>
          </a>
        
        <button type="reset" class="md-search__icon md-icon" title="Clear" aria-label="Clear" tabindex="-1">
          
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"/></svg>
        </button>
      </nav>
      
    </form>
    <div class="md-search__output">
      <div class="md-search__scrollwrap" tabindex="0" data-md-scrollfix>
        <div class="md-search-result" data-md-component="search-result">
          <div class="md-search-result__meta">
            Initializing search
          </div>
          <ol class="md-search-result__list" role="presentation"></ol>
        </div>
      </div>
    </div>
  </div>
</div>
      
    
    
      <div class="md-header__source">
        <a href="https://github.com/genomic-medicine-sweden/nallo" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 496 512"><!--! Font Awesome Free 6.7.2 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2024 Fonticons, Inc.--><path d="M165.9 397.4c0 2-2.3 3.6-5.2 3.6-3.3.3-5.6-1.3-5.6-3.6 0-2 2.3-3.6 5.2-3.6 3-.3 5.6 1.3 5.6 3.6m-31.1-4.5c-.7 2 1.3 4.3 4.3 4.9 2.6 1 5.6 0 6.2-2s-1.3-4.3-4.3-5.2c-2.6-.7-5.5.3-6.2 2.3m44.2-1.7c-2.9.7-4.9 2.6-4.6 4.9.3 2 2.9 3.3 5.9 2.6 2.9-.7 4.9-2.6 4.6-4.6-.3-1.9-3-3.2-5.9-2.9M244.8 8C106.1 8 0 113.3 0 252c0 110.9 69.8 205.8 169.5 239.2 12.8 2.3 17.3-5.6 17.3-12.1 0-6.2-.3-40.4-.3-61.4 0 0-70 15-84.7-29.8 0 0-11.4-29.1-27.8-36.6 0 0-22.9-15.7 1.6-15.4 0 0 24.9 2 38.6 25.8 21.9 38.6 58.6 27.5 72.9 20.9 2.3-16 8.8-27.1 16-33.7-55.9-6.2-112.3-14.3-112.3-110.5 0-27.5 7.6-41.3 23.6-58.9-2.6-6.5-11.1-33.3 2.6-67.9 20.9-6.5 69 27 69 27 20-5.6 41.5-8.5 62.8-8.5s42.8 2.9 62.8 8.5c0 0 48.1-33.6 69-27 13.7 34.7 5.2 61.4 2.6 67.9 16 17.7 25.8 31.5 25.8 58.9 0 96.5-58.9 104.2-114.8 110.5 9.2 7.9 17 22.9 17 46.4 0 33.7-.3 75.4-.3 83.6 0 6.5 4.6 14.4 17.3 12.1C428.2 457.8 496 362.9 496 252 496 113.3 383.5 8 244.8 8M97.2 352.9c-1.3 1-1 3.3.7 5.2 1.6 1.6 3.9 2.3 5.2 1 1.3-1 1-3.3-.7-5.2-1.6-1.6-3.9-2.3-5.2-1m-10.8-8.1c-.7 1.3.3 2.9 2.3 3.9 1.6 1 3.6.7 4.3-.7.7-1.3-.3-2.9-2.3-3.9-2-.6-3.6-.3-4.3.7m32.4 35.6c-1.6 1.3-1 4.3 1.3 6.2 2.3 2.3 5.2 2.6 6.5 1 1.3-1.3.7-4.3-1.3-6.2-2.2-2.3-5.2-2.6-6.5-1m-11.4-14.7c-1.6 1-1.6 3.6 0 5.9s4.3 3.3 5.6 2.3c1.6-1.3 1.6-3.9 0-6.2-1.4-2.3-4-3.3-5.6-2"/></svg>
  </div>
  <div class="md-source__repository">
    genomic-medicine-sweden/nallo
  </div>
</a>
      </div>
    
  </nav>
  
</header>
    
    <div class="md-container" data-md-component="container">
      
      
        
          
        
      
      <main class="md-main" data-md-component="main">
        <div class="md-main__inner md-grid">
          
            
              
              <div class="md-sidebar md-sidebar--primary" data-md-component="sidebar" data-md-type="navigation" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    



<nav class="md-nav md-nav--primary" aria-label="Navigation" data-md-level="0">
  <label class="md-nav__title" for="__drawer">
    <a href=".." title="genomic-medicine-sweden/nallo" class="md-nav__button md-logo" aria-label="genomic-medicine-sweden/nallo" data-md-component="logo">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 16 16"><path d="M2.5 1.75v11.5c0 .138.112.25.25.25h3.17a.75.75 0 0 1 0 1.5H2.75A1.75 1.75 0 0 1 1 13.25V1.75C1 .784 1.784 0 2.75 0h8.5C12.216 0 13 .784 13 1.75v7.736a.75.75 0 0 1-1.5 0V1.75a.25.25 0 0 0-.25-.25h-8.5a.25.25 0 0 0-.25.25m13.274 9.537zl-4.557 4.45a.75.75 0 0 1-1.055-.008l-1.943-1.95a.75.75 0 0 1 1.062-1.058l1.419 1.425 4.026-3.932a.75.75 0 1 1 1.048 1.074M4.75 4h4.5a.75.75 0 0 1 0 1.5h-4.5a.75.75 0 0 1 0-1.5M4 7.75A.75.75 0 0 1 4.75 7h2a.75.75 0 0 1 0 1.5h-2A.75.75 0 0 1 4 7.75"/></svg>

    </a>
    genomic-medicine-sweden/nallo
  </label>
  
    <div class="md-nav__source">
      <a href="https://github.com/genomic-medicine-sweden/nallo" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 496 512"><!--! Font Awesome Free 6.7.2 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2024 Fonticons, Inc.--><path d="M165.9 397.4c0 2-2.3 3.6-5.2 3.6-3.3.3-5.6-1.3-5.6-3.6 0-2 2.3-3.6 5.2-3.6 3-.3 5.6 1.3 5.6 3.6m-31.1-4.5c-.7 2 1.3 4.3 4.3 4.9 2.6 1 5.6 0 6.2-2s-1.3-4.3-4.3-5.2c-2.6-.7-5.5.3-6.2 2.3m44.2-1.7c-2.9.7-4.9 2.6-4.6 4.9.3 2 2.9 3.3 5.9 2.6 2.9-.7 4.9-2.6 4.6-4.6-.3-1.9-3-3.2-5.9-2.9M244.8 8C106.1 8 0 113.3 0 252c0 110.9 69.8 205.8 169.5 239.2 12.8 2.3 17.3-5.6 17.3-12.1 0-6.2-.3-40.4-.3-61.4 0 0-70 15-84.7-29.8 0 0-11.4-29.1-27.8-36.6 0 0-22.9-15.7 1.6-15.4 0 0 24.9 2 38.6 25.8 21.9 38.6 58.6 27.5 72.9 20.9 2.3-16 8.8-27.1 16-33.7-55.9-6.2-112.3-14.3-112.3-110.5 0-27.5 7.6-41.3 23.6-58.9-2.6-6.5-11.1-33.3 2.6-67.9 20.9-6.5 69 27 69 27 20-5.6 41.5-8.5 62.8-8.5s42.8 2.9 62.8 8.5c0 0 48.1-33.6 69-27 13.7 34.7 5.2 61.4 2.6 67.9 16 17.7 25.8 31.5 25.8 58.9 0 96.5-58.9 104.2-114.8 110.5 9.2 7.9 17 22.9 17 46.4 0 33.7-.3 75.4-.3 83.6 0 6.5 4.6 14.4 17.3 12.1C428.2 457.8 496 362.9 496 252 496 113.3 383.5 8 244.8 8M97.2 352.9c-1.3 1-1 3.3.7 5.2 1.6 1.6 3.9 2.3 5.2 1 1.3-1 1-3.3-.7-5.2-1.6-1.6-3.9-2.3-5.2-1m-10.8-8.1c-.7 1.3.3 2.9 2.3 3.9 1.6 1 3.6.7 4.3-.7.7-1.3-.3-2.9-2.3-3.9-2-.6-3.6-.3-4.3.7m32.4 35.6c-1.6 1.3-1 4.3 1.3 6.2 2.3 2.3 5.2 2.6 6.5 1 1.3-1.3.7-4.3-1.3-6.2-2.2-2.3-5.2-2.6-6.5-1m-11.4-14.7c-1.6 1-1.6 3.6 0 5.9s4.3 3.3 5.6 2.3c1.6-1.3 1.6-3.9 0-6.2-1.4-2.3-4-3.3-5.6-2"/></svg>
  </div>
  <div class="md-source__repository">
    genomic-medicine-sweden/nallo
  </div>
</a>
    </div>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href=".." class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Home
    
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../usage/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Usage
    
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
    
  
  
  
    <li class="md-nav__item md-nav__item--active">
      
      <input class="md-nav__toggle md-toggle" type="checkbox" id="__toc">
      
      
        
      
      
        <label class="md-nav__link md-nav__link--active" for="__toc">
          
  
  
  <span class="md-ellipsis">
    Output
    
  </span>
  

          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <a href="./" class="md-nav__link md-nav__link--active">
        
  
  
  <span class="md-ellipsis">
    Output
    
  </span>
  

      </a>
      
        

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#introduction" class="md-nav__link">
    <span class="md-ellipsis">
      Introduction
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#aligned-reads" class="md-nav__link">
    <span class="md-ellipsis">
      Aligned reads
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#assembly" class="md-nav__link">
    <span class="md-ellipsis">
      Assembly
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#methylation-pileups" class="md-nav__link">
    <span class="md-ellipsis">
      Methylation pileups
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#multiqc" class="md-nav__link">
    <span class="md-ellipsis">
      MultiQC
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#pipeline-information" class="md-nav__link">
    <span class="md-ellipsis">
      Pipeline Information
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#phasing" class="md-nav__link">
    <span class="md-ellipsis">
      Phasing
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#qc" class="md-nav__link">
    <span class="md-ellipsis">
      QC
    </span>
  </a>
  
    <nav class="md-nav" aria-label="QC">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#fastqc" class="md-nav__link">
    <span class="md-ellipsis">
      FastQC
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#mosdepth" class="md-nav__link">
    <span class="md-ellipsis">
      Mosdepth
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#cramino" class="md-nav__link">
    <span class="md-ellipsis">
      Cramino
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#somalier" class="md-nav__link">
    <span class="md-ellipsis">
      Somalier
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#peddy" class="md-nav__link">
    <span class="md-ellipsis">
      Peddy
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#deepvariant" class="md-nav__link">
    <span class="md-ellipsis">
      DeepVariant
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#variants" class="md-nav__link">
    <span class="md-ellipsis">
      Variants
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Variants">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#paralogous-genes" class="md-nav__link">
    <span class="md-ellipsis">
      Paralogous genes
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#repeats" class="md-nav__link">
    <span class="md-ellipsis">
      Repeats
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#snvs" class="md-nav__link">
    <span class="md-ellipsis">
      SNVs
    </span>
  </a>
  
    <nav class="md-nav" aria-label="SNVs">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#annotation" class="md-nav__link">
    <span class="md-ellipsis">
      Annotation
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ranking" class="md-nav__link">
    <span class="md-ellipsis">
      Ranking
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#filtering" class="md-nav__link">
    <span class="md-ellipsis">
      Filtering
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
          <li class="md-nav__item">
  <a href="#svs-and-cnvs" class="md-nav__link">
    <span class="md-ellipsis">
      SVs (and CNVs)
    </span>
  </a>
  
    <nav class="md-nav" aria-label="SVs (and CNVs)">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#annotation_1" class="md-nav__link">
    <span class="md-ellipsis">
      Annotation
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ranking_1" class="md-nav__link">
    <span class="md-ellipsis">
      Ranking
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#filtering_1" class="md-nav__link">
    <span class="md-ellipsis">
      Filtering
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#visualization-tracks" class="md-nav__link">
    <span class="md-ellipsis">
      Visualization Tracks
    </span>
  </a>
  
</li>
      
    </ul>
  
</nav>
      
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../parameters/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Parameters
    
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../CITATIONS/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Citations
    
  </span>
  

      </a>
    </li>
  

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
              
              <div class="md-sidebar md-sidebar--secondary" data-md-component="sidebar" data-md-type="toc" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#introduction" class="md-nav__link">
    <span class="md-ellipsis">
      Introduction
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#aligned-reads" class="md-nav__link">
    <span class="md-ellipsis">
      Aligned reads
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#assembly" class="md-nav__link">
    <span class="md-ellipsis">
      Assembly
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#methylation-pileups" class="md-nav__link">
    <span class="md-ellipsis">
      Methylation pileups
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#multiqc" class="md-nav__link">
    <span class="md-ellipsis">
      MultiQC
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#pipeline-information" class="md-nav__link">
    <span class="md-ellipsis">
      Pipeline Information
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#phasing" class="md-nav__link">
    <span class="md-ellipsis">
      Phasing
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#qc" class="md-nav__link">
    <span class="md-ellipsis">
      QC
    </span>
  </a>
  
    <nav class="md-nav" aria-label="QC">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#fastqc" class="md-nav__link">
    <span class="md-ellipsis">
      FastQC
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#mosdepth" class="md-nav__link">
    <span class="md-ellipsis">
      Mosdepth
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#cramino" class="md-nav__link">
    <span class="md-ellipsis">
      Cramino
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#somalier" class="md-nav__link">
    <span class="md-ellipsis">
      Somalier
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#peddy" class="md-nav__link">
    <span class="md-ellipsis">
      Peddy
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#deepvariant" class="md-nav__link">
    <span class="md-ellipsis">
      DeepVariant
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#variants" class="md-nav__link">
    <span class="md-ellipsis">
      Variants
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Variants">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#paralogous-genes" class="md-nav__link">
    <span class="md-ellipsis">
      Paralogous genes
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#repeats" class="md-nav__link">
    <span class="md-ellipsis">
      Repeats
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#snvs" class="md-nav__link">
    <span class="md-ellipsis">
      SNVs
    </span>
  </a>
  
    <nav class="md-nav" aria-label="SNVs">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#annotation" class="md-nav__link">
    <span class="md-ellipsis">
      Annotation
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ranking" class="md-nav__link">
    <span class="md-ellipsis">
      Ranking
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#filtering" class="md-nav__link">
    <span class="md-ellipsis">
      Filtering
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
          <li class="md-nav__item">
  <a href="#svs-and-cnvs" class="md-nav__link">
    <span class="md-ellipsis">
      SVs (and CNVs)
    </span>
  </a>
  
    <nav class="md-nav" aria-label="SVs (and CNVs)">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#annotation_1" class="md-nav__link">
    <span class="md-ellipsis">
      Annotation
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ranking_1" class="md-nav__link">
    <span class="md-ellipsis">
      Ranking
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#filtering_1" class="md-nav__link">
    <span class="md-ellipsis">
      Filtering
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#visualization-tracks" class="md-nav__link">
    <span class="md-ellipsis">
      Visualization Tracks
    </span>
  </a>
  
</li>
      
    </ul>
  
</nav>
                  </div>
                </div>
              </div>
            
          
          
            <div class="md-content" data-md-component="content">
              <article class="md-content__inner md-typeset">
                
                  


  
  


<h1 id="genomic-medicine-swedennallo-output">genomic-medicine-sweden/nallo: Output</h1>
<h2 id="introduction">Introduction</h2>
<p>This document describes the pipeline output files and the tools used to generate them.</p>
<h2 id="aligned-reads">Aligned reads</h2>
<p><a href="https://github.com/lh3/minimap2">Minimap2</a> is used to map the reads to a reference genome. The aligned reads are sorted, merged and indexed using <a href="https://github.com/samtools/samtools">samtools</a>. If the pipeline is run with phasing, the aligned reads will be haplotagged using the active phasing tool.</p>
<table>
<thead>
<tr>
<th>Path</th>
<th>Description</th>
<th>Alignment</th>
<th>Alignment &amp; phasing</th>
</tr>
</thead>
<tbody>
<tr>
<td><code>aligned_reads/minimap2/{sample}/{sample}_aligned.{bam,cram}</code></td>
<td>Alignment file in BAM or CRAM format</td>
<td><img alt="✅" class="twemoji" src="https://cdn.jsdelivr.net/gh/jdecked/twemoji@15.1.0/assets/svg/2705.svg" title=":white_check_mark:" /></td>
<td></td>
</tr>
<tr>
<td><code>aligned_reads/minimap2/{sample}/{sample}_aligned.{bam.bai,cram.crai}</code></td>
<td>Index of the corresponding alignment file</td>
<td><img alt="✅" class="twemoji" src="https://cdn.jsdelivr.net/gh/jdecked/twemoji@15.1.0/assets/svg/2705.svg" title=":white_check_mark:" /></td>
<td></td>
</tr>
</tbody>
</table>
<table>
<thead>
<tr>
<th>Path</th>
<th>Description</th>
<th>Alignment</th>
<th>Alignment &amp; phasing</th>
</tr>
</thead>
<tbody>
<tr>
<td><code>aligned_reads/{sample}/{sample}_haplotagged.{bam,cram}</code></td>
<td>Alignment file with haplotags in BAM or CRAM format</td>
<td></td>
<td><img alt="✅" class="twemoji" src="https://cdn.jsdelivr.net/gh/jdecked/twemoji@15.1.0/assets/svg/2705.svg" title=":white_check_mark:" /></td>
</tr>
<tr>
<td><code>aligned_reads/{sample}/{sample}_haplotagged.{bam.bai,cram.crai}</code></td>
<td>Index of the alignment file</td>
<td></td>
<td><img alt="✅" class="twemoji" src="https://cdn.jsdelivr.net/gh/jdecked/twemoji@15.1.0/assets/svg/2705.svg" title=":white_check_mark:" /></td>
</tr>
</tbody>
</table>
<h2 id="assembly">Assembly</h2>
<p><a href="https://github.com/chhylp123/hifiasm">Hifiasm</a> is used to assemble genomes. The assembled haplotypes are then aligned to the reference genome with <a href="https://github.com/lh3/minimap2">minimap2</a>, tagged with <code>HP:1</code> for the "paternal" haplotype, and <code>HP:2</code> for the "maternal" haplotype, before being merged together into one file with <a href="https://github.com/samtools/samtools">samtools</a>. <a href="https://github.com/vgl-hub/gfastats">gfastats</a> is used to convert the assembly to fasta format before alignment, and also outputs summary stats per haplotype.</p>
<table>
<thead>
<tr>
<th>Path</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr>
<td><code>assembly/sample/{sample}/{sample}_aligned_assembly.{bam,cram}</code></td>
<td>Both assembled haplotypes mapped to the reference genome, merged and haplotagged (<code>HP:1</code>/<code>HP:2</code>).</td>
</tr>
<tr>
<td><code>assembly/sample/{sample}/{sample}_aligned_assembly.{bam.bai,cram.crai}</code></td>
<td>Index of aligned assembly.</td>
</tr>
<tr>
<td><code>assembly/stats/{sample}/{sample}_haplotype_1.assembly_summary</code></td>
<td>Summary statistics for haplotype 1/paternal haplotype</td>
</tr>
<tr>
<td><code>assembly/stats/${sample}/{sample}_haplotype_2.assembly_summary</code></td>
<td>Summary statistics for haplotype 2/maternal haplotype</td>
</tr>
</tbody>
</table>
<h2 id="methylation-pileups">Methylation pileups</h2>
<p><a href="https://github.com/nanoporetech/modkit">Modkit</a> is used to create methylation pileups, producing bedMethyl files for both haplotagged and ungrouped reads. Additionally, methylation information can be viewed in the BAM files, for example in IGV. When phasing is on, modkit outputs pileups per haplotype.</p>
<table>
<thead>
<tr>
<th>Path</th>
<th>Description</th>
<th>Alignment</th>
<th>Alignment &amp; phasing</th>
</tr>
</thead>
<tbody>
<tr>
<td><code>methylation/modkit/pileup/{sample}/*.modkit_pileup_1.bed.gz</code></td>
<td>bedMethyl file with summary counts from haplotagged reads (haplotype 1)</td>
<td></td>
<td><img alt="✅" class="twemoji" src="https://cdn.jsdelivr.net/gh/jdecked/twemoji@15.1.0/assets/svg/2705.svg" title=":white_check_mark:" /></td>
</tr>
<tr>
<td><code>methylation/modkit/pileup/{sample}/*.modkit_pileup_2.bed.gz</code></td>
<td>bedMethyl file with summary counts from haplotagged reads (haplotype 2)</td>
<td></td>
<td><img alt="✅" class="twemoji" src="https://cdn.jsdelivr.net/gh/jdecked/twemoji@15.1.0/assets/svg/2705.svg" title=":white_check_mark:" /></td>
</tr>
<tr>
<td><code>methylation/modkit/pileup/{sample}/*.modkit_pileup_ungrouped.bed.gz</code></td>
<td>bedMethyl file for ungrouped reads</td>
<td></td>
<td><img alt="✅" class="twemoji" src="https://cdn.jsdelivr.net/gh/jdecked/twemoji@15.1.0/assets/svg/2705.svg" title=":white_check_mark:" /></td>
</tr>
<tr>
<td><code>methylation/modkit/pileup/{sample}/*.modkit_pileup.bed.gz</code></td>
<td>bedMethyl file with summary counts from all reads</td>
<td><img alt="✅" class="twemoji" src="https://cdn.jsdelivr.net/gh/jdecked/twemoji@15.1.0/assets/svg/2705.svg" title=":white_check_mark:" /></td>
<td></td>
</tr>
<tr>
<td><code>methylation/modkit/pileup/{sample}/*.bed.gz.tbi</code></td>
<td>Index of the corresponding bedMethyl files</td>
<td><img alt="✅" class="twemoji" src="https://cdn.jsdelivr.net/gh/jdecked/twemoji@15.1.0/assets/svg/2705.svg" title=":white_check_mark:" /></td>
<td></td>
</tr>
</tbody>
</table>
<h2 id="multiqc">MultiQC</h2>
<p><a href="http://multiqc.info">MultiQC</a> generates an HTML report summarizing all samples' QC results and pipeline statistics.</p>
<table>
<thead>
<tr>
<th>Path</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr>
<td><code>multiqc/multiqc_report.html</code></td>
<td>HTML report summarizing QC results</td>
</tr>
<tr>
<td><code>multiqc/multiqc_data/</code></td>
<td>Directory containing parsed statistics</td>
</tr>
<tr>
<td><code>multiqc/multiqc_plots/</code></td>
<td>Directory containing static report images</td>
</tr>
</tbody>
</table>
<h2 id="pipeline-information">Pipeline Information</h2>
<p><a href="https://www.nextflow.io/docs/latest/tracing.html">Nextflow</a> generates reports for troubleshooting, performance, and traceability.</p>
<table>
<thead>
<tr>
<th>Path</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr>
<td><code>pipeline_info/execution_report.html</code></td>
<td>Execution report</td>
</tr>
<tr>
<td><code>pipeline_info/execution_timeline.html</code></td>
<td>Timeline report</td>
</tr>
<tr>
<td><code>pipeline_info/execution_trace.txt</code></td>
<td>Execution trace</td>
</tr>
<tr>
<td><code>pipeline_info/pipeline_dag.dot</code></td>
<td>Pipeline DAG in DOT format</td>
</tr>
<tr>
<td><code>pipeline_info/pipeline_report.html</code></td>
<td>Pipeline report</td>
</tr>
<tr>
<td><code>pipeline_info/software_versions.yml</code></td>
<td>Software versions used in the run</td>
</tr>
</tbody>
</table>
<h2 id="phasing">Phasing</h2>
<p><a href="https://github.com/twolinin/longphase">LongPhase</a>, <a href="https://whatshap.readthedocs.io/en/latest/">WhatsHap</a>, or <a href="https://github.com/PacificBiosciences/HiPhase">HiPhase</a> are used for phasing.</p>
<table>
<thead>
<tr>
<th>Path</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr>
<td><code>aligned_reads/{sample}/{sample}_haplotagged.{bam,cram}</code></td>
<td>BAM/CRAM file with haplotags</td>
</tr>
<tr>
<td><code>aligned_reads/{sample}/{sample}_haplotagged.{bam.bai,cram.crai}</code></td>
<td>Index of the BAM/CRAM file</td>
</tr>
<tr>
<td><code>phased_variants/{sample}/*.vcf.gz</code></td>
<td>VCF file with phased variants</td>
</tr>
<tr>
<td><code>phased_variants/{sample}/*.vcf.gz.tbi</code></td>
<td>Index of the VCF file</td>
</tr>
<tr>
<td><code>qc/phasing_stats/{sample}/*.blocks.tsv.gz</code></td>
<td>Phase block file</td>
</tr>
<tr>
<td><code>qc/phasing_stats/{sample}/*.blocks.tsv.gz.tbi</code></td>
<td>Index of block file</td>
</tr>
<tr>
<td><code>qc/phasing_stats/{sample}/*.stats.tsv</code></td>
<td>Phasing statistics file</td>
</tr>
</tbody>
</table>
<h2 id="qc">QC</h2>
<p><a href="http://www.bioinformatics.babraham.ac.uk/projects/fastqc/">FastQC</a>, <a href="https://github.com/wdecoster/cramino">cramino</a>, <a href="https://github.com/brentp/mosdepth">mosdepth</a>, and <a href="https://github.com/brentp/somalier">somalier</a> are used for read quality control.</p>
<h3 id="fastqc">FastQC</h3>
<p><a href="http://www.bioinformatics.babraham.ac.uk/projects/fastqc/">FastQC</a> provides general quality metrics for sequenced reads, including information on quality score distribution, per-base sequence content (%A/T/G/C), adapter contamination, and overrepresented sequences. For more details, refer to the <a href="http://www.bioinformatics.babraham.ac.uk/projects/fastqc/Help/">FastQC help pages</a>.</p>
<table>
<thead>
<tr>
<th>Path</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr>
<td><code>qc/fastqc/{sample}/*_fastqc.html</code></td>
<td>FastQC report containing quality metrics</td>
</tr>
<tr>
<td><code>qc/fastqc/{sample}/*_fastqc.zip</code></td>
<td>Zip archive with the FastQC report, data files, and plot images</td>
</tr>
</tbody>
</table>
<h3 id="mosdepth">Mosdepth</h3>
<p><a href="https://github.com/brentp/mosdepth">Mosdepth</a> is used to report quality control metrics such as coverage and GC content from alignment files.</p>
<table>
<thead>
<tr>
<th>Path</th>
<th>Description</th>
<th>With <code>--target_regions</code></th>
<th>Without <code>--target_regions</code></th>
</tr>
</thead>
<tbody>
<tr>
<td><code>qc/mosdepth/{sample}/{sample}.mosdepth.global.dist.txt</code></td>
<td>Cumulative distribution of bases covered for at least a given coverage value, across chromosomes and the whole genome</td>
<td><img alt="✅" class="twemoji" src="https://cdn.jsdelivr.net/gh/jdecked/twemoji@15.1.0/assets/svg/2705.svg" title=":white_check_mark:" /></td>
<td><img alt="✅" class="twemoji" src="https://cdn.jsdelivr.net/gh/jdecked/twemoji@15.1.0/assets/svg/2705.svg" title=":white_check_mark:" /></td>
</tr>
<tr>
<td><code>qc/mosdepth/{sample}/{sample}.mosdepth.summary.txt</code></td>
<td>Mosdepth summary file</td>
<td><img alt="✅" class="twemoji" src="https://cdn.jsdelivr.net/gh/jdecked/twemoji@15.1.0/assets/svg/2705.svg" title=":white_check_mark:" /></td>
<td><img alt="✅" class="twemoji" src="https://cdn.jsdelivr.net/gh/jdecked/twemoji@15.1.0/assets/svg/2705.svg" title=":white_check_mark:" /></td>
</tr>
<tr>
<td><code>qc/mosdepth/{sample}/{sample}.mosdepth.region.dist.txt</code></td>
<td>Cumulative distribution of bases covered for at least a given coverage value, across regions</td>
<td><img alt="✅" class="twemoji" src="https://cdn.jsdelivr.net/gh/jdecked/twemoji@15.1.0/assets/svg/2705.svg" title=":white_check_mark:" /></td>
<td></td>
</tr>
<tr>
<td><code>qc/mosdepth/{sample}/{sample}.per-base.d4</code></td>
<td>Per-base depth in d4 format</td>
<td><img alt="✅" class="twemoji" src="https://cdn.jsdelivr.net/gh/jdecked/twemoji@15.1.0/assets/svg/2705.svg" title=":white_check_mark:" /></td>
<td></td>
</tr>
<tr>
<td><code>qc/mosdepth/{sample}/{sample}.regions.bed.gz</code></td>
<td>Depth per region</td>
<td><img alt="✅" class="twemoji" src="https://cdn.jsdelivr.net/gh/jdecked/twemoji@15.1.0/assets/svg/2705.svg" title=":white_check_mark:" /></td>
<td></td>
</tr>
<tr>
<td><code>qc/mosdepth/{sample}/{sample}.regions.bed.gz.csi</code></td>
<td>Index of the regions.bed.gz file</td>
<td><img alt="✅" class="twemoji" src="https://cdn.jsdelivr.net/gh/jdecked/twemoji@15.1.0/assets/svg/2705.svg" title=":white_check_mark:" /></td>
<td></td>
</tr>
</tbody>
</table>
<h3 id="cramino">Cramino</h3>
<p><a href="https://github.com/wdecoster/cramino">cramino</a> is used to analyze both phased and unphased reads.</p>
<table>
<thead>
<tr>
<th>Path</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr>
<td><code>qc/cramino/phased/{sample}/*.arrow</code></td>
<td>Read length and quality in <a href="https://arrow.apache.org/docs/format/Columnar.html">Apache Arrow</a> format</td>
</tr>
<tr>
<td><code>qc/cramino/phased/{sample}/*.txt</code></td>
<td>Summary information in text format</td>
</tr>
<tr>
<td><code>qc/cramino/unphased/{sample}/*.arrow</code></td>
<td>Read length and quality in <a href="https://arrow.apache.org/docs/format/Columnar.html">Apache Arrow</a> format</td>
</tr>
<tr>
<td><code>qc/cramino/unphased/{sample}/*.txt</code></td>
<td>Summary information in text format</td>
</tr>
</tbody>
</table>
<h3 id="somalier">Somalier</h3>
<p><a href="https://github.com/brentp/somalier">somalier</a> checks relatedness and sex.</p>
<table>
<thead>
<tr>
<th>Path</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr>
<td><code>pedigree/family/{family).ped</code></td>
<td>PED file updated with somalier-inferred sex per family</td>
</tr>
<tr>
<td><code>qc/somalier/relate/{project}/{project}.html</code></td>
<td>HTML report</td>
</tr>
<tr>
<td><code>qc/somalier/relate/{project}/{project}.pairs.tsv</code></td>
<td>Information about sample pairs</td>
</tr>
<tr>
<td><code>qc/somalier/relate/{project}/{project}.samples.tsv</code></td>
<td>Information about individual samples</td>
</tr>
</tbody>
</table>
<h3 id="peddy">Peddy</h3>
<p><a href="https://github.com/brentp/peddy">peddy</a> checks relatedness and sex.</p>
<table>
<thead>
<tr>
<th>Path</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr>
<td><code>qc/peddy/{family}/{family}.peddy.ped</code></td>
<td>PED file updated with peddy-inferred sex per family</td>
</tr>
<tr>
<td><code>qc/peddy/{family}/{family}.html</code></td>
<td>HTML report</td>
</tr>
<tr>
<td><code>qc/peddy/{family}/{family}.vs.html</code></td>
<td>HTML report of observed vs expected relatedness</td>
</tr>
<tr>
<td><code>qc/peddy/{family}/{family}.sex_check.csv</code></td>
<td>Comparison between reported sex (ped file) and that inferred from peddy</td>
</tr>
<tr>
<td><code>qc/peddy/{family}/{family}.het_check.csv</code></td>
<td>Het check does general QC including rate of het calls, allele-balance at het calls, mean and median depth, and a PCA projection onto thousand genomes. Incudes ancestry check</td>
</tr>
<tr>
<td><code>qc/peddy/{family}/{family}.ped_check.csv</code></td>
<td>Ped check compares the relatedness of 2 samples as reported in a .ped file to the relatedness inferred from the genotypes and ~25K sites in the genome</td>
</tr>
<tr>
<td><code>qc/peddy/{family}/{family}.sex_check.png</code></td>
<td>PNG comparison between reported sex (ped file) and that inferred from peddy</td>
</tr>
<tr>
<td><code>qc/peddy/{family}/{family}.het_check.png</code></td>
<td>PNG of heterozygosity check</td>
</tr>
<tr>
<td><code>qc/peddy/{family}/{family}.ped_check.png</code></td>
<td>PNG of the ped check comparison</td>
</tr>
<tr>
<td><code>qc/peddy/{family}/{family}.ped_check.rel-difference.csv</code></td>
<td>CSV file with the comparison between inferred and given relatedness</td>
</tr>
</tbody>
</table>
<h3 id="deepvariant">DeepVariant</h3>
<p><code>vcf_stats_report.py</code> from <a href="https://github.com/google/deepvariant">DeepVariant</a> is used to generate a html report per sample.</p>
<table>
<thead>
<tr>
<th>Path</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr>
<td><code>qc/deepvariant_vcfstatsreport/{sample}/${sample}.visual_report.html</code></td>
<td>Visual report of SNV calls from DeepVariant</td>
</tr>
</tbody>
</table>
<h2 id="variants">Variants</h2>
<p>In general, annotated variant calls are output per family while unannotated calls are output per sample.</p>
<h3 id="paralogous-genes">Paralogous genes</h3>
<p><a href="https://github.com/PacificBiosciences/paraphase">Paraphase</a> is used to call paralogous genes.</p>
<table>
<thead>
<tr>
<th>Path</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr>
<td><code>paraphase/sample/{sample}/{sample}.paraphase.{bam,cram}</code></td>
<td>BAM/CRAM file with reads from analyzed regions</td>
</tr>
<tr>
<td><code>paraphase/sample/{sample}/{sample}.paraphase.{bam.bai,cram.crai}</code></td>
<td>Index of the BAM/CRAM file</td>
</tr>
<tr>
<td><code>paraphase/sample/{sample}/{sample}.paraphase.json</code></td>
<td>Summary of haplotypes and variant calls</td>
</tr>
<tr>
<td><code>paraphase/sample/{sample}_paraphase_vcfs/{sample}_{gene}_vcf.gz</code></td>
<td>VCF file per gene</td>
</tr>
<tr>
<td><code>paraphase/sample/{sample}_paraphase_vcfs/{sample}_{gene}_vcf.gz.tbi</code></td>
<td>Index of the VCF file</td>
</tr>
<tr>
<td><code>paraphase/family/{family_id}/{family_id}_paraphase_merged.vcf.gz</code></td>
<td>VCF file from paraphase, merged by family</td>
</tr>
<tr>
<td><code>paraphase/family/{family_id}/{family_id}_paraphase_merged.vcf.gz.tbi</code></td>
<td>Index of the VCF file merged by family</td>
</tr>
<tr>
<td><code>paraphase/family/{family_id}/{family_id}_merged.json</code></td>
<td>Summary of haplotypes and variant calls, merged by family</td>
</tr>
</tbody>
</table>
<h3 id="repeats">Repeats</h3>
<p><a href="https://github.com/PacificBiosciences/trgt">TRGT</a> or <a href="https://github.com/wdecoster/STRdust">STRdust</a> are used to call repeats.
<a href="https://github.com/Clinical-Genomics/stranger">Stranger</a> is used to annotate repeats.</p>
<table>
<thead>
<tr>
<th>Path</th>
<th>Description</th>
<th>STRdust, Call repeats only</th>
<th>TRGT, Call repeats only</th>
<th>TRGT, Call &amp; annotate repeats</th>
</tr>
</thead>
<tbody>
<tr>
<td><code>repeats/sample/{sample}/{sample}_{str_caller}.vcf.gz</code></td>
<td>VCF file with called repeats for a sample</td>
<td><img alt="✅" class="twemoji" src="https://cdn.jsdelivr.net/gh/jdecked/twemoji@15.1.0/assets/svg/2705.svg" title=":white_check_mark:" /></td>
<td><img alt="✅" class="twemoji" src="https://cdn.jsdelivr.net/gh/jdecked/twemoji@15.1.0/assets/svg/2705.svg" title=":white_check_mark:" /></td>
<td><img alt="✅" class="twemoji" src="https://cdn.jsdelivr.net/gh/jdecked/twemoji@15.1.0/assets/svg/2705.svg" title=":white_check_mark:" /></td>
</tr>
<tr>
<td><code>repeats/sample/{sample}/{sample}_{str_caller}.vcf.gz.tbi</code></td>
<td>Index of the VCF file</td>
<td><img alt="✅" class="twemoji" src="https://cdn.jsdelivr.net/gh/jdecked/twemoji@15.1.0/assets/svg/2705.svg" title=":white_check_mark:" /></td>
<td><img alt="✅" class="twemoji" src="https://cdn.jsdelivr.net/gh/jdecked/twemoji@15.1.0/assets/svg/2705.svg" title=":white_check_mark:" /></td>
<td><img alt="✅" class="twemoji" src="https://cdn.jsdelivr.net/gh/jdecked/twemoji@15.1.0/assets/svg/2705.svg" title=":white_check_mark:" /></td>
</tr>
<tr>
<td><code>repeats/sample/{sample}/{sample}_spanning_trgt.{bam,cram}</code></td>
<td>BAM/CRAM file with sorted spanning reads</td>
<td></td>
<td><img alt="✅" class="twemoji" src="https://cdn.jsdelivr.net/gh/jdecked/twemoji@15.1.0/assets/svg/2705.svg" title=":white_check_mark:" /></td>
<td><img alt="✅" class="twemoji" src="https://cdn.jsdelivr.net/gh/jdecked/twemoji@15.1.0/assets/svg/2705.svg" title=":white_check_mark:" /></td>
</tr>
<tr>
<td><code>repeats/sample/{sample}/{sample}_spanning_trgt.{bam.bai,cram.crai}</code></td>
<td>Index of the BAM/CRAM file</td>
<td></td>
<td><img alt="✅" class="twemoji" src="https://cdn.jsdelivr.net/gh/jdecked/twemoji@15.1.0/assets/svg/2705.svg" title=":white_check_mark:" /></td>
<td><img alt="✅" class="twemoji" src="https://cdn.jsdelivr.net/gh/jdecked/twemoji@15.1.0/assets/svg/2705.svg" title=":white_check_mark:" /></td>
</tr>
<tr>
<td><code>repeats/family/{family}/{family}_repeat_expansions.vcf.gz</code></td>
<td>Merged VCF file per family</td>
<td><img alt="✅" class="twemoji" src="https://cdn.jsdelivr.net/gh/jdecked/twemoji@15.1.0/assets/svg/2705.svg" title=":white_check_mark:" /></td>
<td><img alt="✅" class="twemoji" src="https://cdn.jsdelivr.net/gh/jdecked/twemoji@15.1.0/assets/svg/2705.svg" title=":white_check_mark:" /></td>
<td></td>
</tr>
<tr>
<td><code>repeats/family/{family}/{family}_repeat_expansions.vcf.gz.tbi</code></td>
<td>Index of the VCF file</td>
<td><img alt="✅" class="twemoji" src="https://cdn.jsdelivr.net/gh/jdecked/twemoji@15.1.0/assets/svg/2705.svg" title=":white_check_mark:" /></td>
<td><img alt="✅" class="twemoji" src="https://cdn.jsdelivr.net/gh/jdecked/twemoji@15.1.0/assets/svg/2705.svg" title=":white_check_mark:" /></td>
<td></td>
</tr>
<tr>
<td><code>repeats/family/{family}/{family}_repeat_expansions_annotated.vcf.gz</code></td>
<td>Merged, annotated VCF file per family</td>
<td></td>
<td></td>
<td><img alt="✅" class="twemoji" src="https://cdn.jsdelivr.net/gh/jdecked/twemoji@15.1.0/assets/svg/2705.svg" title=":white_check_mark:" /></td>
</tr>
<tr>
<td><code>repeats/family/{family}/{family}_repeat_expansions_annotated.vcf.gz.tbi</code></td>
<td>Index of the VCF file</td>
<td></td>
<td></td>
<td><img alt="✅" class="twemoji" src="https://cdn.jsdelivr.net/gh/jdecked/twemoji@15.1.0/assets/svg/2705.svg" title=":white_check_mark:" /></td>
</tr>
</tbody>
</table>
<h3 id="snvs">SNVs</h3>
<p><a href="https://github.com/google/deepvariant">DeepVariant</a> is used to call variants, while <a href="https://samtools.github.io/bcftools/bcftools.html">bcftools</a> and <a href="https://github.com/dnanexus-rnd/GLnexus">GLnexus</a> are used for merging variants.</p>
<table>
<thead>
<tr>
<th>Path</th>
<th>Description</th>
<th>Call SNVs</th>
<th>Call &amp; annotate SNVs</th>
<th>Call, annotate and rank SNVs</th>
</tr>
</thead>
<tbody>
<tr>
<td><code>snvs/sample/{sample}/{sample}_snvs.vcf.gz</code></td>
<td>VCF file containing called variants with alternative genotypes for a sample</td>
<td><img alt="✅" class="twemoji" src="https://cdn.jsdelivr.net/gh/jdecked/twemoji@15.1.0/assets/svg/2705.svg" title=":white_check_mark:" /></td>
<td><img alt="✅" class="twemoji" src="https://cdn.jsdelivr.net/gh/jdecked/twemoji@15.1.0/assets/svg/2705.svg" title=":white_check_mark:" /></td>
<td><img alt="✅" class="twemoji" src="https://cdn.jsdelivr.net/gh/jdecked/twemoji@15.1.0/assets/svg/2705.svg" title=":white_check_mark:" /></td>
</tr>
<tr>
<td><code>snvs/sample/{sample}/{sample}_snvs.vcf.gz.tbi</code></td>
<td>Index of the corresponding VCF file</td>
<td><img alt="✅" class="twemoji" src="https://cdn.jsdelivr.net/gh/jdecked/twemoji@15.1.0/assets/svg/2705.svg" title=":white_check_mark:" /></td>
<td><img alt="✅" class="twemoji" src="https://cdn.jsdelivr.net/gh/jdecked/twemoji@15.1.0/assets/svg/2705.svg" title=":white_check_mark:" /></td>
<td><img alt="✅" class="twemoji" src="https://cdn.jsdelivr.net/gh/jdecked/twemoji@15.1.0/assets/svg/2705.svg" title=":white_check_mark:" /></td>
</tr>
<tr>
<td><code>snvs/stats/sample/*.stats.txt</code></td>
<td>Variant statistics</td>
<td><img alt="✅" class="twemoji" src="https://cdn.jsdelivr.net/gh/jdecked/twemoji@15.1.0/assets/svg/2705.svg" title=":white_check_mark:" /></td>
<td><img alt="✅" class="twemoji" src="https://cdn.jsdelivr.net/gh/jdecked/twemoji@15.1.0/assets/svg/2705.svg" title=":white_check_mark:" /></td>
<td><img alt="✅" class="twemoji" src="https://cdn.jsdelivr.net/gh/jdecked/twemoji@15.1.0/assets/svg/2705.svg" title=":white_check_mark:" /></td>
</tr>
<tr>
<td><code>qc/deepvariant_vcfstatsreport/{sample}/${sample}.visual_report.html</code></td>
<td>Visual report of SNV calls from DeepVariant</td>
<td><img alt="✅" class="twemoji" src="https://cdn.jsdelivr.net/gh/jdecked/twemoji@15.1.0/assets/svg/2705.svg" title=":white_check_mark:" /></td>
<td><img alt="✅" class="twemoji" src="https://cdn.jsdelivr.net/gh/jdecked/twemoji@15.1.0/assets/svg/2705.svg" title=":white_check_mark:" /></td>
<td><img alt="✅" class="twemoji" src="https://cdn.jsdelivr.net/gh/jdecked/twemoji@15.1.0/assets/svg/2705.svg" title=":white_check_mark:" /></td>
</tr>
<tr>
<td><code>snvs/family/{family}/{family}_snvs.vcf.gz</code></td>
<td>VCF file containing called variants for all samples</td>
<td><img alt="✅" class="twemoji" src="https://cdn.jsdelivr.net/gh/jdecked/twemoji@15.1.0/assets/svg/2705.svg" title=":white_check_mark:" /></td>
<td></td>
<td></td>
</tr>
<tr>
<td><code>snvs/family/{family}/{family}_snvs.vcf.gz.tbi</code></td>
<td>Index of the corresponding VCF file</td>
<td><img alt="✅" class="twemoji" src="https://cdn.jsdelivr.net/gh/jdecked/twemoji@15.1.0/assets/svg/2705.svg" title=":white_check_mark:" /></td>
<td></td>
<td></td>
</tr>
</tbody>
</table>
<h4 id="annotation">Annotation</h4>
<p><a href="https://github.com/brentp/echtvar">Echtvar</a> and <a href="https://www.ensembl.org/vep">VEP</a> are used for annotating SNVs, while <a href="https://cadd.gs.washington.edu/">CADD</a> is used to annotate INDELs with CADD scores.</p>
<table>
<thead>
<tr>
<th>Path</th>
<th>Description</th>
<th>Call SNVs</th>
<th>Call &amp; annotate SNVs</th>
<th>Call, annotate and rank SNVs</th>
</tr>
</thead>
<tbody>
<tr>
<td><code>snvs/sample/{sample}/{sample}_snvs_annotated.vcf.gz</code></td>
<td>VCF file containing annotated variants with alternative genotypes for a sample</td>
<td></td>
<td><img alt="✅" class="twemoji" src="https://cdn.jsdelivr.net/gh/jdecked/twemoji@15.1.0/assets/svg/2705.svg" title=":white_check_mark:" /></td>
<td></td>
</tr>
<tr>
<td><code>snvs/sample/{sample}/{sample}_snvs_annotated.vcf.gz.tbi</code></td>
<td>Index of the annotated VCF file</td>
<td></td>
<td><img alt="✅" class="twemoji" src="https://cdn.jsdelivr.net/gh/jdecked/twemoji@15.1.0/assets/svg/2705.svg" title=":white_check_mark:" /></td>
<td></td>
</tr>
<tr>
<td><code>snvs/family/{family}/{family}_snvs_annotated.vcf.gz</code></td>
<td>VCF file containing annotated variants per family</td>
<td></td>
<td><img alt="✅" class="twemoji" src="https://cdn.jsdelivr.net/gh/jdecked/twemoji@15.1.0/assets/svg/2705.svg" title=":white_check_mark:" /></td>
<td></td>
</tr>
<tr>
<td><code>snvs/family/{family}/{family}_snvs_annotated.vcf.gz.tbi</code></td>
<td>Index of the annotated VCF file</td>
<td></td>
<td><img alt="✅" class="twemoji" src="https://cdn.jsdelivr.net/gh/jdecked/twemoji@15.1.0/assets/svg/2705.svg" title=":white_check_mark:" /></td>
<td></td>
</tr>
</tbody>
</table>
<h4 id="ranking">Ranking</h4>
<p><a href="https://github.com/Clinical-Genomics/genmod">GENMOD</a> is used to rank the annotated SNVs and INDELs.</p>
<table>
<thead>
<tr>
<th>Path</th>
<th>Description</th>
<th>Call SNVs</th>
<th>Call &amp; annotate SNVs</th>
<th>Call, annotate and rank SNVs</th>
</tr>
</thead>
<tbody>
<tr>
<td><code>snvs/sample/{sample}/{sample}_snvs_annotated_ranked.vcf.gz</code></td>
<td>VCF file with annotated and ranked variants for a sample</td>
<td></td>
<td><img alt="✅" class="twemoji" src="https://cdn.jsdelivr.net/gh/jdecked/twemoji@15.1.0/assets/svg/2705.svg" title=":white_check_mark:" /></td>
<td></td>
</tr>
<tr>
<td><code>snvs/sample/{sample}/{sample}_snvs_annotated_ranked.vcf.gz.tbi</code></td>
<td>Index of the ranked VCF file</td>
<td></td>
<td><img alt="✅" class="twemoji" src="https://cdn.jsdelivr.net/gh/jdecked/twemoji@15.1.0/assets/svg/2705.svg" title=":white_check_mark:" /></td>
<td></td>
</tr>
<tr>
<td><code>snvs/family/{family}/{family}_snvs_annotated_ranked.vcf.gz</code></td>
<td>VCF file with annotated and ranked variants per family</td>
<td></td>
<td></td>
<td><img alt="✅" class="twemoji" src="https://cdn.jsdelivr.net/gh/jdecked/twemoji@15.1.0/assets/svg/2705.svg" title=":white_check_mark:" /></td>
</tr>
<tr>
<td><code>snvs/family/{family}/{family}_snvs_annotated_ranked.vcf.gz.tbi</code></td>
<td>Index of the ranked VCF file</td>
<td></td>
<td></td>
<td><img alt="✅" class="twemoji" src="https://cdn.jsdelivr.net/gh/jdecked/twemoji@15.1.0/assets/svg/2705.svg" title=":white_check_mark:" /></td>
</tr>
</tbody>
</table>
<h4 id="filtering">Filtering</h4>
<p><a href="https://www.ensembl.org/vep">Filter_vep</a> and <a href="https://samtools.github.io/bcftools/bcftools.html">bcftools</a> can be used to filter variants. These will be output if either of <code>--filter_variants_hgnc_id</code> and <code>--filter_snvs_expression</code> has been used, and only family VCFs are filtered.</p>
<table>
<thead>
<tr>
<th>Path</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr>
<td><code>snvs/{family}/{family}_*_filtered.vcf.gz</code></td>
<td>VCF file with filtered variants for a family</td>
</tr>
<tr>
<td><code>snvs/{family}/{family}_*_filtered.vcf.gz.tbi</code></td>
<td>Index of the filtered VCF file</td>
</tr>
</tbody>
</table>
<div class="admonition tip">
<p class="admonition-title">Tip</p>
<p>Filtered variants are output alongside unfiltered variants as additional files.</p>
</div>
<h3 id="svs-and-cnvs">SVs (and CNVs)</h3>
<p><a href="https://github.com/KolmogorovLab/Severus">Severus</a> or <a href="https://github.com/fritzsedlazeck/Sniffles">Sniffles</a> are used to call structural variants, while <a href="https://github.com/PacificBiosciences/HiFiCNV">HiFiCNV</a> is used to call CNVs. HiFiCNV also produces copy number, depth, and MAF <a href="#visualization-tracks">visualization tracks</a>.</p>
<div class="admonition info">
<p class="admonition-title">Variant merging strategies</p>
<p>SV and CNV calls are output unmerged per sample, while the family files are first merged between samples for SVs and CNVs separately, then the merged SV and CNV files are merged again, with priority given to coordinates from the SV calls. SV calls are output for all callers, but only variants from one caller (set by <code>--sv_caller</code>) are merged with CNVs, then annotated, ranked and filtered.</p>
</div>
<table>
<thead>
<tr>
<th>Path</th>
<th>Description</th>
<th>Call SVs</th>
<th>Call CNVs</th>
<th>Call SVs &amp; CNVs</th>
<th><code>--publish_unannotated_family_svs</code></th>
</tr>
</thead>
<tbody>
<tr>
<td><code>svs/sample/{sample}/{sample}_{sniffles,severus}_svs_merged.vcf.gz</code></td>
<td>VCF file with SVs per sample</td>
<td><img alt="✅" class="twemoji" src="https://cdn.jsdelivr.net/gh/jdecked/twemoji@15.1.0/assets/svg/2705.svg" title=":white_check_mark:" /></td>
<td></td>
<td><img alt="✅" class="twemoji" src="https://cdn.jsdelivr.net/gh/jdecked/twemoji@15.1.0/assets/svg/2705.svg" title=":white_check_mark:" /></td>
<td></td>
</tr>
<tr>
<td><code>svs/sample/{sample}/{sample}_{sniffles,severus}_svs_merged.vcf.gz.tbi</code></td>
<td>VCF file with SVs per sample</td>
<td><img alt="✅" class="twemoji" src="https://cdn.jsdelivr.net/gh/jdecked/twemoji@15.1.0/assets/svg/2705.svg" title=":white_check_mark:" /></td>
<td></td>
<td><img alt="✅" class="twemoji" src="https://cdn.jsdelivr.net/gh/jdecked/twemoji@15.1.0/assets/svg/2705.svg" title=":white_check_mark:" /></td>
<td></td>
</tr>
<tr>
<td><code>svs/sample/{sample}/{sample}_hificnv_cnvs_merged.vcf.gz</code></td>
<td>VCF file with CNVs per sample</td>
<td></td>
<td><img alt="✅" class="twemoji" src="https://cdn.jsdelivr.net/gh/jdecked/twemoji@15.1.0/assets/svg/2705.svg" title=":white_check_mark:" /></td>
<td><img alt="✅" class="twemoji" src="https://cdn.jsdelivr.net/gh/jdecked/twemoji@15.1.0/assets/svg/2705.svg" title=":white_check_mark:" /></td>
<td></td>
</tr>
<tr>
<td><code>svs/sample/{sample}/{sample}_hificnv_cnvs_merged.vcf.gz.tbi</code></td>
<td>VCF file with CNVs per sample</td>
<td></td>
<td><img alt="✅" class="twemoji" src="https://cdn.jsdelivr.net/gh/jdecked/twemoji@15.1.0/assets/svg/2705.svg" title=":white_check_mark:" /></td>
<td><img alt="✅" class="twemoji" src="https://cdn.jsdelivr.net/gh/jdecked/twemoji@15.1.0/assets/svg/2705.svg" title=":white_check_mark:" /></td>
<td></td>
</tr>
<tr>
<td><code>svs/family/{family_id}/{family_id}_${hifiasm,sniffles,severus}_{svs,cnvs}_merged.vcf.gz</code></td>
<td>VCF file with merged SVs per family and caller</td>
<td></td>
<td></td>
<td></td>
<td><img alt="✅" class="twemoji" src="https://cdn.jsdelivr.net/gh/jdecked/twemoji@15.1.0/assets/svg/2705.svg" title=":white_check_mark:" /></td>
</tr>
<tr>
<td><code>svs/family/{family_id}/{family_id}_${hifiasm,sniffles,severus}_{snvs,cnvs}_merged.vcf.gz.tbi</code></td>
<td>Index of the merged VCF file</td>
<td></td>
<td></td>
<td></td>
<td><img alt="✅" class="twemoji" src="https://cdn.jsdelivr.net/gh/jdecked/twemoji@15.1.0/assets/svg/2705.svg" title=":white_check_mark:" /></td>
</tr>
<tr>
<td><code>svs/family/{family_id}/{family_id}_cnvs_svs_merged.vcf.gz</code></td>
<td>VCF file with merged CNVs and SVs per family</td>
<td></td>
<td></td>
<td><img alt="✅" class="twemoji" src="https://cdn.jsdelivr.net/gh/jdecked/twemoji@15.1.0/assets/svg/2705.svg" title=":white_check_mark:" /></td>
<td></td>
</tr>
<tr>
<td><code>svs/family/{family_id}/{family_id}_cnvs_svs_merged.vcf.gz.tbi</code></td>
<td>Index of the merged VCF file</td>
<td></td>
<td></td>
<td><img alt="✅" class="twemoji" src="https://cdn.jsdelivr.net/gh/jdecked/twemoji@15.1.0/assets/svg/2705.svg" title=":white_check_mark:" /></td>
<td></td>
</tr>
</tbody>
</table>
<h4 id="annotation_1">Annotation</h4>
<p><a href="https://github.com/J35P312/SVDB">SVDB</a> and <a href="https://www.ensembl.org/vep">VEP</a> are used to annotate structural variants.</p>
<table>
<thead>
<tr>
<th>Path</th>
<th>Description</th>
<th>Call &amp; annotate SVs</th>
<th> Call &amp; annotate SVs &amp; CNVs</th>
</tr>
</thead>
<tbody>
<tr>
<td><code>svs/family/{family_id}/{family_id}_cnvs_svs_merged_annotated.vcf.gz</code></td>
<td>VCF file with merged and annotated CNVs and SVs per family</td>
<td></td>
<td><img alt="✅" class="twemoji" src="https://cdn.jsdelivr.net/gh/jdecked/twemoji@15.1.0/assets/svg/2705.svg" title=":white_check_mark:" /></td>
</tr>
<tr>
<td><code>svs/family/{family_id}/{family_id}_cnvs_svs_merged_annotated.vcf.gz.tbi</code></td>
<td>Index of the merged VCF file</td>
<td></td>
<td><img alt="✅" class="twemoji" src="https://cdn.jsdelivr.net/gh/jdecked/twemoji@15.1.0/assets/svg/2705.svg" title=":white_check_mark:" /></td>
</tr>
<tr>
<td><code>svs/family/{family_id}/{family_id}_svs_merged_annotated.vcf.gz</code></td>
<td>VCF file with merged and annotated SVs per family</td>
<td><img alt="✅" class="twemoji" src="https://cdn.jsdelivr.net/gh/jdecked/twemoji@15.1.0/assets/svg/2705.svg" title=":white_check_mark:" /></td>
<td></td>
</tr>
<tr>
<td><code>svs/family/{family_id}/{family_id}_svs_merged_annotated.vcf.gz.tbi</code></td>
<td>Index of the merged VCF file</td>
<td><img alt="✅" class="twemoji" src="https://cdn.jsdelivr.net/gh/jdecked/twemoji@15.1.0/assets/svg/2705.svg" title=":white_check_mark:" /></td>
<td></td>
</tr>
</tbody>
</table>
<h4 id="ranking_1">Ranking</h4>
<p><a href="https://github.com/Clinical-Genomics/genmod">GENMOD</a> is used to rank the annotated SVs.</p>
<table>
<thead>
<tr>
<th>Path</th>
<th>Description</th>
<th>Rank SVs</th>
<th>Rank SVs &amp; CNVs</th>
</tr>
</thead>
<tbody>
<tr>
<td><code>svs/family/{family_id}/{family_id}_cnvs_svs_merged_annotated_ranked.vcf.gz</code></td>
<td>VCF file with merged, annotated and ranked CNVs and SVs per family</td>
<td></td>
<td><img alt="✅" class="twemoji" src="https://cdn.jsdelivr.net/gh/jdecked/twemoji@15.1.0/assets/svg/2705.svg" title=":white_check_mark:" /></td>
</tr>
<tr>
<td><code>svs/family/{family_id}/{family_id}_cnvs_svs_merged_annotated_ranked.vcf.gz.tbi</code></td>
<td>Index of the merged VCF file</td>
<td></td>
<td><img alt="✅" class="twemoji" src="https://cdn.jsdelivr.net/gh/jdecked/twemoji@15.1.0/assets/svg/2705.svg" title=":white_check_mark:" /></td>
</tr>
<tr>
<td><code>svs/family/{family_id}/{family_id}_svs_merged_annotated_ranked.vcf.gz</code></td>
<td>VCF file with merged, annotated and ranked SVs per family</td>
<td><img alt="✅" class="twemoji" src="https://cdn.jsdelivr.net/gh/jdecked/twemoji@15.1.0/assets/svg/2705.svg" title=":white_check_mark:" /></td>
<td></td>
</tr>
<tr>
<td><code>svs/family/{family_id}/{family_id}_svs_merged_annotated_ranked.vcf.gz.tbi</code></td>
<td>Index of the merged VCF file</td>
<td><img alt="✅" class="twemoji" src="https://cdn.jsdelivr.net/gh/jdecked/twemoji@15.1.0/assets/svg/2705.svg" title=":white_check_mark:" /></td>
<td></td>
</tr>
</tbody>
</table>
<h4 id="filtering_1">Filtering</h4>
<p><a href="https://www.ensembl.org/vep">Filter_vep</a> and <a href="https://samtools.github.io/bcftools/bcftools.html">bcftools</a> can be used to filter variants. These will be output if either of <code>--filter_variants_hgnc_id</code> and <code>--filter_svs_expression</code> has been used, and only family VCFs are filtered.</p>
<table>
<thead>
<tr>
<th>Path</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr>
<td><code>svs/{family}/{family}_*_filtered.vcf.gz</code></td>
<td>VCF file with filtered variants for a family</td>
</tr>
<tr>
<td><code>svs/{family}/{family}_*_filtered.vcf.gz.tbi</code></td>
<td>Index of the filtered VCF file</td>
</tr>
</tbody>
</table>
<div class="admonition tip">
<p class="admonition-title">Tip</p>
<p>Filtered variants are output alongside unfiltered variants as additional files.</p>
</div>
<h2 id="visualization-tracks">Visualization Tracks</h2>
<p><a href="https://github.com/PacificBiosciences/HiFiCNV">HiFiCNV</a> is used to call CNVs, but it also produces copy number, depth, and MAF tracks that can be visualized in for example IGV.</p>
<table>
<thead>
<tr>
<th>Path</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr>
<td><code>visualization_tracks/{sample}/*.copynum.bedgraph</code></td>
<td>Copy number in bedgraph format</td>
</tr>
<tr>
<td><code>visualization_tracks/{sample}/*.depth.bw</code></td>
<td>Depth track in BigWig format</td>
</tr>
<tr>
<td><code>visualization_tracks/{sample}/*.maf.bw</code></td>
<td>Minor allele frequencies in BigWig format</td>
</tr>
</tbody>
</table>












                
              </article>
            </div>
          
          
<script>var target=document.getElementById(location.hash.slice(1));target&&target.name&&(target.checked=target.name.startsWith("__tabbed_"))</script>
        </div>
        
          <button type="button" class="md-top md-icon" data-md-component="top" hidden>
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M13 20h-2V8l-5.5 5.5-1.42-1.42L12 4.16l7.92 7.92-1.42 1.42L13 8z"/></svg>
  Back to top
</button>
        
      </main>
      
        <footer class="md-footer">
  
    
      
      <nav class="md-footer__inner md-grid" aria-label="Footer" >
        
          
          <a href="../usage/" class="md-footer__link md-footer__link--prev" aria-label="Previous: Usage">
            <div class="md-footer__button md-icon">
              
              <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11z"/></svg>
            </div>
            <div class="md-footer__title">
              <span class="md-footer__direction">
                Previous
              </span>
              <div class="md-ellipsis">
                Usage
              </div>
            </div>
          </a>
        
        
          
          <a href="../parameters/" class="md-footer__link md-footer__link--next" aria-label="Next: Parameters">
            <div class="md-footer__title">
              <span class="md-footer__direction">
                Next
              </span>
              <div class="md-ellipsis">
                Parameters
              </div>
            </div>
            <div class="md-footer__button md-icon">
              
              <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M4 11v2h12l-5.5 5.5 1.42 1.42L19.84 12l-7.92-7.92L10.5 5.5 16 11z"/></svg>
            </div>
          </a>
        
      </nav>
    
  
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-copyright">
  
  
    Made with
    <a href="https://squidfunk.github.io/mkdocs-material/" target="_blank" rel="noopener">
      Material for MkDocs
    </a>
  
</div>
      
    </div>
  </div>
</footer>
      
    </div>
    <div class="md-dialog" data-md-component="dialog">
      <div class="md-dialog__inner md-typeset"></div>
    </div>
    
    
    
      
      <script id="__config" type="application/json">{"base": "..", "features": ["content.code.annotate", "content.code.copy", "navigation.instant", "navigation.top", "navigation.tracking", "navigation.sections", "navigation.footer", "search.share", "toc.follow"], "search": "../assets/javascripts/workers/search.f8cc74c7.min.js", "tags": null, "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version": "Select version"}, "version": {"provider": "mike"}}</script>
    
    
      <script src="../assets/javascripts/bundle.c8b220af.min.js"></script>
      
    
  </body>
</html>