
<!doctype html>
<html lang="en" class="no-js">
  <head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      
      
      
        <link rel="canonical" href="https://genomic-medicine-sweden.github.io/nallo/dev/parameters/">
      
      
        <link rel="prev" href="../output/">
      
      
        <link rel="next" href="../CITATIONS/">
      
      
      <link rel="icon" href="../assets/images/favicon.png">
      <meta name="generator" content="mkdocs-1.6.1, mkdocs-material-9.6.13">
    
    
      
        <title>Parameters - genomic-medicine-sweden/nallo</title>
      
    
    
      <link rel="stylesheet" href="../assets/stylesheets/main.342714a4.min.css">
      
        
        <link rel="stylesheet" href="../assets/stylesheets/palette.06af60db.min.css">
      
      


    
    
      
    
    
      
        
        
        <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Inter:300,300i,400,400i,700,700i%7CRoboto+Mono:400,400i,700,700i&display=fallback">
        <style>:root{--md-text-font:"Inter";--md-code-font:"Roboto Mono"}</style>
      
    
    
    <script>__md_scope=new URL("..",location),__md_hash=e=>[...e].reduce(((e,_)=>(e<<5)-e+_.charCodeAt(0)),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script>
    
      

    
    
    
  </head>
  
  
    
    
      
    
    
    
    
    <body dir="ltr" data-md-color-scheme="slate" data-md-color-primary="custom" data-md-color-accent="custom">
  
    
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off">
    <label class="md-overlay" for="__drawer"></label>
    <div data-md-component="skip">
      
        
        <a href="#genomic-medicine-swedennallo-pipeline-parameters" class="md-skip">
          Skip to content
        </a>
      
    </div>
    <div data-md-component="announce">
      
    </div>
    
      <div data-md-color-scheme="default" data-md-component="outdated" hidden>
        
      </div>
    
    
      

  

<header class="md-header md-header--shadow" data-md-component="header">
  <nav class="md-header__inner md-grid" aria-label="Header">
    <a href=".." title="genomic-medicine-sweden/nallo" class="md-header__button md-logo" aria-label="genomic-medicine-sweden/nallo" data-md-component="logo">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 16 16"><path d="M2.5 1.75v11.5c0 .138.112.25.25.25h3.17a.75.75 0 0 1 0 1.5H2.75A1.75 1.75 0 0 1 1 13.25V1.75C1 .784 1.784 0 2.75 0h8.5C12.216 0 13 .784 13 1.75v7.736a.75.75 0 0 1-1.5 0V1.75a.25.25 0 0 0-.25-.25h-8.5a.25.25 0 0 0-.25.25m13.274 9.537zl-4.557 4.45a.75.75 0 0 1-1.055-.008l-1.943-1.95a.75.75 0 0 1 1.062-1.058l1.419 1.425 4.026-3.932a.75.75 0 1 1 1.048 1.074M4.75 4h4.5a.75.75 0 0 1 0 1.5h-4.5a.75.75 0 0 1 0-1.5M4 7.75A.75.75 0 0 1 4.75 7h2a.75.75 0 0 1 0 1.5h-2A.75.75 0 0 1 4 7.75"/></svg>

    </a>
    <label class="md-header__button md-icon" for="__drawer">
      
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M3 6h18v2H3zm0 5h18v2H3zm0 5h18v2H3z"/></svg>
    </label>
    <div class="md-header__title" data-md-component="header-title">
      <div class="md-header__ellipsis">
        <div class="md-header__topic">
          <span class="md-ellipsis">
            genomic-medicine-sweden/nallo
          </span>
        </div>
        <div class="md-header__topic" data-md-component="header-topic">
          <span class="md-ellipsis">
            
              Parameters
            
          </span>
        </div>
      </div>
    </div>
    
      
        <form class="md-header__option" data-md-component="palette">
  
    
    
    
    <input class="md-option" data-md-color-media="(prefers-color-scheme: dark)" data-md-color-scheme="slate" data-md-color-primary="custom" data-md-color-accent="custom"  aria-label="Switch to light mode"  type="radio" name="__palette" id="__palette_0">
    
      <label class="md-header__button md-icon" title="Switch to light mode" for="__palette_1" hidden>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 7a5 5 0 0 1 5 5 5 5 0 0 1-5 5 5 5 0 0 1-5-5 5 5 0 0 1 5-5m0 2a3 3 0 0 0-3 3 3 3 0 0 0 3 3 3 3 0 0 0 3-3 3 3 0 0 0-3-3m0-7 2.39 3.42C13.65 5.15 12.84 5 12 5s-1.65.15-2.39.42zM3.34 7l4.16-.35A7.2 7.2 0 0 0 5.94 8.5c-.44.74-.69 1.5-.83 2.29zm.02 10 1.76-3.77a7.131 7.131 0 0 0 2.38 4.14zM20.65 7l-1.77 3.79a7.02 7.02 0 0 0-2.38-4.15zm-.01 10-4.14.36c.59-.51 1.12-1.14 1.54-1.86.42-.73.69-1.5.83-2.29zM12 22l-2.41-3.44c.74.27 1.55.44 2.41.44.82 0 1.63-.17 2.37-.44z"/></svg>
      </label>
    
  
    
    
    
    <input class="md-option" data-md-color-media="(prefers-color-scheme: light)" data-md-color-scheme="default" data-md-color-primary="custom" data-md-color-accent="custom"  aria-label="Switch to dark mode"  type="radio" name="__palette" id="__palette_1">
    
      <label class="md-header__button md-icon" title="Switch to dark mode" for="__palette_0" hidden>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="m17.75 4.09-2.53 1.94.91 3.06-2.63-1.81-2.63 1.81.91-3.06-2.53-1.94L12.44 4l1.06-3 1.06 3zm3.5 6.91-1.64 1.25.59 1.98-1.7-1.17-1.7 1.17.59-1.98L15.75 11l2.06-.05L18.5 9l.69 1.95zm-2.28 4.95c.83-.08 1.72 1.1 1.19 1.85-.32.45-.66.87-1.08 1.27C15.17 23 8.84 23 4.94 19.07c-3.91-3.9-3.91-10.24 0-14.14.4-.4.82-.76 1.27-1.08.75-.53 1.93.36 1.85 1.19-.27 2.86.69 5.83 2.89 8.02a9.96 9.96 0 0 0 8.02 2.89m-1.64 2.02a12.08 12.08 0 0 1-7.8-3.47c-2.17-2.19-3.33-5-3.49-7.82-2.81 3.14-2.7 7.96.31 10.98 3.02 3.01 7.84 3.12 10.98.31"/></svg>
      </label>
    
  
</form>
      
    
    
      <script>var palette=__md_get("__palette");if(palette&&palette.color){if("(prefers-color-scheme)"===palette.color.media){var media=matchMedia("(prefers-color-scheme: light)"),input=document.querySelector(media.matches?"[data-md-color-media='(prefers-color-scheme: light)']":"[data-md-color-media='(prefers-color-scheme: dark)']");palette.color.media=input.getAttribute("data-md-color-media"),palette.color.scheme=input.getAttribute("data-md-color-scheme"),palette.color.primary=input.getAttribute("data-md-color-primary"),palette.color.accent=input.getAttribute("data-md-color-accent")}for(var[key,value]of Object.entries(palette.color))document.body.setAttribute("data-md-color-"+key,value)}</script>
    
    
    
      
      
        <label class="md-header__button md-icon" for="__search">
          
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"/></svg>
        </label>
        <div class="md-search" data-md-component="search" role="dialog">
  <label class="md-search__overlay" for="__search"></label>
  <div class="md-search__inner" role="search">
    <form class="md-search__form" name="search">
      <input type="text" class="md-search__input" name="query" aria-label="Search" placeholder="Search" autocapitalize="off" autocorrect="off" autocomplete="off" spellcheck="false" data-md-component="search-query" required>
      <label class="md-search__icon md-icon" for="__search">
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"/></svg>
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11z"/></svg>
      </label>
      <nav class="md-search__options" aria-label="Search">
        
          <a href="javascript:void(0)" class="md-search__icon md-icon" title="Share" aria-label="Share" data-clipboard data-clipboard-text="" data-md-component="search-share" tabindex="-1">
            
            <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M18 16.08c-.76 0-1.44.3-1.96.77L8.91 12.7c.05-.23.09-.46.09-.7s-.04-.47-.09-.7l7.05-4.11c.54.5 1.25.81 2.04.81a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3c0 .24.04.47.09.7L8.04 9.81C7.5 9.31 6.79 9 6 9a3 3 0 0 0-3 3 3 3 0 0 0 3 3c.79 0 1.5-.31 2.04-.81l7.12 4.15c-.05.21-.08.43-.08.66 0 1.61 1.31 2.91 2.92 2.91s2.92-1.3 2.92-2.91A2.92 2.92 0 0 0 18 16.08"/></svg>
          </a>
        
        <button type="reset" class="md-search__icon md-icon" title="Clear" aria-label="Clear" tabindex="-1">
          
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"/></svg>
        </button>
      </nav>
      
    </form>
    <div class="md-search__output">
      <div class="md-search__scrollwrap" tabindex="0" data-md-scrollfix>
        <div class="md-search-result" data-md-component="search-result">
          <div class="md-search-result__meta">
            Initializing search
          </div>
          <ol class="md-search-result__list" role="presentation"></ol>
        </div>
      </div>
    </div>
  </div>
</div>
      
    
    
      <div class="md-header__source">
        <a href="https://github.com/genomic-medicine-sweden/nallo" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 496 512"><!--! Font Awesome Free 6.7.2 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2024 Fonticons, Inc.--><path d="M165.9 397.4c0 2-2.3 3.6-5.2 3.6-3.3.3-5.6-1.3-5.6-3.6 0-2 2.3-3.6 5.2-3.6 3-.3 5.6 1.3 5.6 3.6m-31.1-4.5c-.7 2 1.3 4.3 4.3 4.9 2.6 1 5.6 0 6.2-2s-1.3-4.3-4.3-5.2c-2.6-.7-5.5.3-6.2 2.3m44.2-1.7c-2.9.7-4.9 2.6-4.6 4.9.3 2 2.9 3.3 5.9 2.6 2.9-.7 4.9-2.6 4.6-4.6-.3-1.9-3-3.2-5.9-2.9M244.8 8C106.1 8 0 113.3 0 252c0 110.9 69.8 205.8 169.5 239.2 12.8 2.3 17.3-5.6 17.3-12.1 0-6.2-.3-40.4-.3-61.4 0 0-70 15-84.7-29.8 0 0-11.4-29.1-27.8-36.6 0 0-22.9-15.7 1.6-15.4 0 0 24.9 2 38.6 25.8 21.9 38.6 58.6 27.5 72.9 20.9 2.3-16 8.8-27.1 16-33.7-55.9-6.2-112.3-14.3-112.3-110.5 0-27.5 7.6-41.3 23.6-58.9-2.6-6.5-11.1-33.3 2.6-67.9 20.9-6.5 69 27 69 27 20-5.6 41.5-8.5 62.8-8.5s42.8 2.9 62.8 8.5c0 0 48.1-33.6 69-27 13.7 34.7 5.2 61.4 2.6 67.9 16 17.7 25.8 31.5 25.8 58.9 0 96.5-58.9 104.2-114.8 110.5 9.2 7.9 17 22.9 17 46.4 0 33.7-.3 75.4-.3 83.6 0 6.5 4.6 14.4 17.3 12.1C428.2 457.8 496 362.9 496 252 496 113.3 383.5 8 244.8 8M97.2 352.9c-1.3 1-1 3.3.7 5.2 1.6 1.6 3.9 2.3 5.2 1 1.3-1 1-3.3-.7-5.2-1.6-1.6-3.9-2.3-5.2-1m-10.8-8.1c-.7 1.3.3 2.9 2.3 3.9 1.6 1 3.6.7 4.3-.7.7-1.3-.3-2.9-2.3-3.9-2-.6-3.6-.3-4.3.7m32.4 35.6c-1.6 1.3-1 4.3 1.3 6.2 2.3 2.3 5.2 2.6 6.5 1 1.3-1.3.7-4.3-1.3-6.2-2.2-2.3-5.2-2.6-6.5-1m-11.4-14.7c-1.6 1-1.6 3.6 0 5.9s4.3 3.3 5.6 2.3c1.6-1.3 1.6-3.9 0-6.2-1.4-2.3-4-3.3-5.6-2"/></svg>
  </div>
  <div class="md-source__repository">
    genomic-medicine-sweden/nallo
  </div>
</a>
      </div>
    
  </nav>
  
</header>
    
    <div class="md-container" data-md-component="container">
      
      
        
          
        
      
      <main class="md-main" data-md-component="main">
        <div class="md-main__inner md-grid">
          
            
              
              <div class="md-sidebar md-sidebar--primary" data-md-component="sidebar" data-md-type="navigation" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    



<nav class="md-nav md-nav--primary" aria-label="Navigation" data-md-level="0">
  <label class="md-nav__title" for="__drawer">
    <a href=".." title="genomic-medicine-sweden/nallo" class="md-nav__button md-logo" aria-label="genomic-medicine-sweden/nallo" data-md-component="logo">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 16 16"><path d="M2.5 1.75v11.5c0 .138.112.25.25.25h3.17a.75.75 0 0 1 0 1.5H2.75A1.75 1.75 0 0 1 1 13.25V1.75C1 .784 1.784 0 2.75 0h8.5C12.216 0 13 .784 13 1.75v7.736a.75.75 0 0 1-1.5 0V1.75a.25.25 0 0 0-.25-.25h-8.5a.25.25 0 0 0-.25.25m13.274 9.537zl-4.557 4.45a.75.75 0 0 1-1.055-.008l-1.943-1.95a.75.75 0 0 1 1.062-1.058l1.419 1.425 4.026-3.932a.75.75 0 1 1 1.048 1.074M4.75 4h4.5a.75.75 0 0 1 0 1.5h-4.5a.75.75 0 0 1 0-1.5M4 7.75A.75.75 0 0 1 4.75 7h2a.75.75 0 0 1 0 1.5h-2A.75.75 0 0 1 4 7.75"/></svg>

    </a>
    genomic-medicine-sweden/nallo
  </label>
  
    <div class="md-nav__source">
      <a href="https://github.com/genomic-medicine-sweden/nallo" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 496 512"><!--! Font Awesome Free 6.7.2 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2024 Fonticons, Inc.--><path d="M165.9 397.4c0 2-2.3 3.6-5.2 3.6-3.3.3-5.6-1.3-5.6-3.6 0-2 2.3-3.6 5.2-3.6 3-.3 5.6 1.3 5.6 3.6m-31.1-4.5c-.7 2 1.3 4.3 4.3 4.9 2.6 1 5.6 0 6.2-2s-1.3-4.3-4.3-5.2c-2.6-.7-5.5.3-6.2 2.3m44.2-1.7c-2.9.7-4.9 2.6-4.6 4.9.3 2 2.9 3.3 5.9 2.6 2.9-.7 4.9-2.6 4.6-4.6-.3-1.9-3-3.2-5.9-2.9M244.8 8C106.1 8 0 113.3 0 252c0 110.9 69.8 205.8 169.5 239.2 12.8 2.3 17.3-5.6 17.3-12.1 0-6.2-.3-40.4-.3-61.4 0 0-70 15-84.7-29.8 0 0-11.4-29.1-27.8-36.6 0 0-22.9-15.7 1.6-15.4 0 0 24.9 2 38.6 25.8 21.9 38.6 58.6 27.5 72.9 20.9 2.3-16 8.8-27.1 16-33.7-55.9-6.2-112.3-14.3-112.3-110.5 0-27.5 7.6-41.3 23.6-58.9-2.6-6.5-11.1-33.3 2.6-67.9 20.9-6.5 69 27 69 27 20-5.6 41.5-8.5 62.8-8.5s42.8 2.9 62.8 8.5c0 0 48.1-33.6 69-27 13.7 34.7 5.2 61.4 2.6 67.9 16 17.7 25.8 31.5 25.8 58.9 0 96.5-58.9 104.2-114.8 110.5 9.2 7.9 17 22.9 17 46.4 0 33.7-.3 75.4-.3 83.6 0 6.5 4.6 14.4 17.3 12.1C428.2 457.8 496 362.9 496 252 496 113.3 383.5 8 244.8 8M97.2 352.9c-1.3 1-1 3.3.7 5.2 1.6 1.6 3.9 2.3 5.2 1 1.3-1 1-3.3-.7-5.2-1.6-1.6-3.9-2.3-5.2-1m-10.8-8.1c-.7 1.3.3 2.9 2.3 3.9 1.6 1 3.6.7 4.3-.7.7-1.3-.3-2.9-2.3-3.9-2-.6-3.6-.3-4.3.7m32.4 35.6c-1.6 1.3-1 4.3 1.3 6.2 2.3 2.3 5.2 2.6 6.5 1 1.3-1.3.7-4.3-1.3-6.2-2.2-2.3-5.2-2.6-6.5-1m-11.4-14.7c-1.6 1-1.6 3.6 0 5.9s4.3 3.3 5.6 2.3c1.6-1.3 1.6-3.9 0-6.2-1.4-2.3-4-3.3-5.6-2"/></svg>
  </div>
  <div class="md-source__repository">
    genomic-medicine-sweden/nallo
  </div>
</a>
    </div>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href=".." class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Home
    
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../usage/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Usage
    
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../output/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Output
    
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
    
  
  
  
    <li class="md-nav__item md-nav__item--active">
      
      <input class="md-nav__toggle md-toggle" type="checkbox" id="__toc">
      
      
        
      
      
        <label class="md-nav__link md-nav__link--active" for="__toc">
          
  
  
  <span class="md-ellipsis">
    Parameters
    
  </span>
  

          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <a href="./" class="md-nav__link md-nav__link--active">
        
  
  
  <span class="md-ellipsis">
    Parameters
    
  </span>
  

      </a>
      
        

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#workflow-skip-options" class="md-nav__link">
    <span class="md-ellipsis">
      Workflow skip options
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#inputoutput-options" class="md-nav__link">
    <span class="md-ellipsis">
      Input/output options
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#reference-genome-options" class="md-nav__link">
    <span class="md-ellipsis">
      Reference genome options
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#institutional-config-options" class="md-nav__link">
    <span class="md-ellipsis">
      Institutional config options
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#generic-options" class="md-nav__link">
    <span class="md-ellipsis">
      Generic options
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#workflow-options" class="md-nav__link">
    <span class="md-ellipsis">
      Workflow options
    </span>
  </a>
  
</li>
      
    </ul>
  
</nav>
      
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../CITATIONS/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Citations
    
  </span>
  

      </a>
    </li>
  

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
              
              <div class="md-sidebar md-sidebar--secondary" data-md-component="sidebar" data-md-type="toc" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#workflow-skip-options" class="md-nav__link">
    <span class="md-ellipsis">
      Workflow skip options
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#inputoutput-options" class="md-nav__link">
    <span class="md-ellipsis">
      Input/output options
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#reference-genome-options" class="md-nav__link">
    <span class="md-ellipsis">
      Reference genome options
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#institutional-config-options" class="md-nav__link">
    <span class="md-ellipsis">
      Institutional config options
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#generic-options" class="md-nav__link">
    <span class="md-ellipsis">
      Generic options
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#workflow-options" class="md-nav__link">
    <span class="md-ellipsis">
      Workflow options
    </span>
  </a>
  
</li>
      
    </ul>
  
</nav>
                  </div>
                </div>
              </div>
            
          
          
            <div class="md-content" data-md-component="content">
              <article class="md-content__inner md-typeset">
                
                  


  
  


<h1 id="genomic-medicine-swedennallo-pipeline-parameters">genomic-medicine-sweden/nallo pipeline parameters</h1>
<p>Long-read variant calling pipeline</p>
<h2 id="workflow-skip-options">Workflow skip options</h2>
<p>Allows skipping certain parts of the pipeline</p>
<table>
<thead>
<tr>
<th>Parameter</th>
<th>Description</th>
<th>Type</th>
<th>Default</th>
<th>Required</th>
<th>Hidden</th>
</tr>
</thead>
<tbody>
<tr>
<td><code>skip_qc</code></td>
<td>Skip QC of reads</td>
<td><code>boolean</code></td>
<td>False</td>
<td></td>
<td></td>
</tr>
<tr>
<td><code>skip_snv_calling</code></td>
<td>Skip short variant calling</td>
<td><code>boolean</code></td>
<td>False</td>
<td></td>
<td></td>
</tr>
<tr>
<td><code>skip_genome_assembly</code></td>
<td>Skip genome assembly and assembly variant calling</td>
<td><code>boolean</code></td>
<td>False</td>
<td></td>
<td></td>
</tr>
<tr>
<td><code>skip_alignment</code></td>
<td>Skip read mapping (alignment)</td>
<td><code>boolean</code></td>
<td>False</td>
<td></td>
<td></td>
</tr>
<tr>
<td><code>skip_methylation_pileups</code></td>
<td>Skip generation of methylation pileups</td>
<td><code>boolean</code></td>
<td>False</td>
<td></td>
<td></td>
</tr>
<tr>
<td><code>skip_repeat_calling</code></td>
<td>Skip tandem repeat calling</td>
<td><code>boolean</code></td>
<td>False</td>
<td></td>
<td></td>
</tr>
<tr>
<td><code>skip_repeat_annotation</code></td>
<td>Skip tandem repeat annotation</td>
<td><code>boolean</code></td>
<td>False</td>
<td></td>
<td></td>
</tr>
<tr>
<td><code>skip_peddy</code></td>
<td>Skip peddy</td>
<td><code>boolean</code></td>
<td>False</td>
<td></td>
<td></td>
</tr>
<tr>
<td><code>skip_phasing</code></td>
<td>Skip phasing of variants and haplotagging of reads</td>
<td><code>boolean</code></td>
<td>False</td>
<td></td>
<td></td>
</tr>
<tr>
<td><code>skip_snv_annotation</code></td>
<td>Skip short variant annotation</td>
<td><code>boolean</code></td>
<td>False</td>
<td></td>
<td></td>
</tr>
<tr>
<td><code>skip_sv_calling</code></td>
<td>Skip structural variant calling</td>
<td><code>boolean</code></td>
<td>False</td>
<td></td>
<td></td>
</tr>
<tr>
<td><code>skip_sv_annotation</code></td>
<td>Skip structural variant annotation</td>
<td><code>boolean</code></td>
<td>False</td>
<td></td>
<td></td>
</tr>
<tr>
<td><code>skip_cnv_calling</code></td>
<td>Skip CNV calling</td>
<td><code>boolean</code></td>
<td>False</td>
<td></td>
<td></td>
</tr>
<tr>
<td><code>skip_call_paralogs</code></td>
<td>Skip the calling of specific paralogous genes</td>
<td><code>boolean</code></td>
<td>False</td>
<td></td>
<td></td>
</tr>
<tr>
<td><code>skip_rank_variants</code></td>
<td>Skip ranking of short variants</td>
<td><code>boolean</code></td>
<td>False</td>
<td></td>
<td></td>
</tr>
</tbody>
</table>
<h2 id="inputoutput-options">Input/output options</h2>
<p>Define where the pipeline should find input data and save output data.</p>
<table>
<thead>
<tr>
<th>Parameter</th>
<th>Description</th>
<th>Type</th>
<th>Default</th>
<th>Required</th>
<th>Hidden</th>
</tr>
</thead>
<tbody>
<tr>
<td><code>input</code></td>
<td>Path to comma-separated file containing information about the samples in the experiment. <details><summary>Help</summary><small>You will need to create a design file with information about the samples in your experiment before running the pipeline. Use this parameter to specify its location. It has to be a comma-separated file with 3 columns, and a header row.</small></details></td>
<td><code>string</code></td>
<td></td>
<td>True</td>
<td></td>
</tr>
<tr>
<td><code>outdir</code></td>
<td>The output directory where the results will be saved. You have to use absolute paths to storage on Cloud infrastructure.</td>
<td><code>string</code></td>
<td></td>
<td>True</td>
<td></td>
</tr>
<tr>
<td><code>email</code></td>
<td>Email address for completion summary. <details><summary>Help</summary><small>Set this parameter to your e-mail address to get a summary e-mail with details of the run sent to you when the workflow exits. If set in your user config file (<code>~/.nextflow/config</code>) then you don't need to specify this on the command line for every run.</small></details></td>
<td><code>string</code></td>
<td></td>
<td></td>
<td></td>
</tr>
<tr>
<td><code>multiqc_title</code></td>
<td>MultiQC report title. Printed as page header, used for filename if not otherwise specified.</td>
<td><code>string</code></td>
<td></td>
<td></td>
<td></td>
</tr>
<tr>
<td><code>cadd_prescored_indels</code></td>
<td>Path to a directory containing prescored indels for CADD. <details><summary>Help</summary><small>This folder contains the compressed files and indexes that would otherwise be in data/prescored folder as described in https://github.com/kircherlab/CADD-scripts/#manual-installation.</small></details></td>
<td><code>string</code></td>
<td></td>
<td></td>
<td></td>
</tr>
<tr>
<td><code>cadd_resources</code></td>
<td>Path to a directory containing CADD annotations. <details><summary>Help</summary><small>This folder contains the uncompressed files that would otherwise be in data/annotation folder as described in https://github.com/kircherlab/CADD-scripts/#manual-installation.</small></details></td>
<td><code>string</code></td>
<td></td>
<td></td>
<td></td>
</tr>
<tr>
<td><code>par_regions</code></td>
<td>Provide a bed file of chrX and chrY PAR regions for DeepVariant</td>
<td><code>string</code></td>
<td></td>
<td></td>
<td></td>
</tr>
<tr>
<td><code>tandem_repeats</code></td>
<td>A tandem repeat BED file for sniffles</td>
<td><code>string</code></td>
<td></td>
<td></td>
<td></td>
</tr>
<tr>
<td><code>str_bed</code></td>
<td>A BED file with repeats to be genotyped with TRGT</td>
<td><code>string</code></td>
<td></td>
<td></td>
<td></td>
</tr>
<tr>
<td><code>echtvar_snv_databases</code></td>
<td>A csv file with echtvar databases to annotate SNVs with</td>
<td><code>string</code></td>
<td></td>
<td></td>
<td></td>
</tr>
<tr>
<td><code>svdb_sv_databases</code></td>
<td>Databases used for structural variant annotation in vcf format. <details><summary>Help</summary><small>Path to comma-separated file containing information about the databases used for structural variant annotation.</small></details></td>
<td><code>string</code></td>
<td></td>
<td></td>
<td></td>
</tr>
<tr>
<td><code>stranger_repeat_catalog</code></td>
<td>A variant catalog json-file for stranger</td>
<td><code>string</code></td>
<td></td>
<td></td>
<td></td>
</tr>
<tr>
<td><code>variant_consequences_snvs</code></td>
<td>File containing list of SO terms listed in the order of severity from most severe to lease severe for annotating genomic SNVs. For more information check https://ensembl.org/info/genome/variation/prediction/predicted_data.html</td>
<td><code>string</code></td>
<td></td>
<td></td>
<td></td>
</tr>
<tr>
<td><code>variant_consequences_svs</code></td>
<td>File containing list of SO terms listed in the order of severity from most severe to lease severe for annotating genomic SVs. For more information check https://ensembl.org/info/genome/variation/prediction/predicted_data.html</td>
<td><code>string</code></td>
<td></td>
<td></td>
<td></td>
</tr>
<tr>
<td><code>vep_cache</code></td>
<td>A path to the VEP cache location</td>
<td><code>string</code></td>
<td></td>
<td></td>
<td></td>
</tr>
<tr>
<td><code>target_regions</code></td>
<td>A BED file with regions of interest, used to limit variant calling.</td>
<td><code>string</code></td>
<td></td>
<td></td>
<td></td>
</tr>
<tr>
<td><code>hificnv_expected_xy_cn</code></td>
<td>A BED file containing expected copy number regions for XY samples.</td>
<td><code>string</code></td>
<td></td>
<td></td>
<td></td>
</tr>
<tr>
<td><code>hificnv_expected_xx_cn</code></td>
<td>A BED file containing expected copy number regions for XX samples.</td>
<td><code>string</code></td>
<td></td>
<td></td>
<td></td>
</tr>
<tr>
<td><code>hificnv_excluded_regions</code></td>
<td>A BED file specifying regions to exclude with HiFiCNV, such as centromeres.</td>
<td><code>string</code></td>
<td></td>
<td></td>
<td></td>
</tr>
<tr>
<td><code>genmod_reduced_penetrance</code></td>
<td>A file with gene ids that have reduced penetrance. For use with genmod.</td>
<td><code>string</code></td>
<td></td>
<td></td>
<td></td>
</tr>
<tr>
<td><code>genmod_score_config_snvs</code></td>
<td>A SNV rank model config file for genmod.</td>
<td><code>string</code></td>
<td></td>
<td></td>
<td></td>
</tr>
<tr>
<td><code>genmod_score_config_svs</code></td>
<td>A SV rank model config file for genmod.</td>
<td><code>string</code></td>
<td></td>
<td></td>
<td></td>
</tr>
<tr>
<td><code>somalier_sites</code></td>
<td>A VCF of known polymorphic sites for somalier</td>
<td><code>string</code></td>
<td></td>
<td></td>
<td></td>
</tr>
<tr>
<td><code>alignment_output_format</code></td>
<td>Output format for alignment files. Either <code>bam</code> or <code>cram</code></td>
<td><code>string</code></td>
<td>bam</td>
<td></td>
<td></td>
</tr>
<tr>
<td><code>peddy_sites</code></td>
<td>A file path to a VCF of known polymorphic sites for peddy. You may need to create a custom sites file if you have incomplete or targeted data.</td>
<td><code>string</code></td>
<td></td>
<td></td>
<td></td>
</tr>
<tr>
<td><code>modules_testdata_base_path</code></td>
<td>Base URL or local path to location of modules test dataset files</td>
<td><code>string</code></td>
<td>https://raw.githubusercontent.com/nf-core/test-datasets/modules/data/</td>
<td></td>
<td>True</td>
</tr>
<tr>
<td><code>pipelines_testdata_base_path</code></td>
<td>Base URL or local path to location of pipeline test dataset files</td>
<td><code>string</code></td>
<td>https://raw.githubusercontent.com/genomic-medicine-sweden/test-datasets/c7e99d0b2a366c606a5b3e5626b36e727c794a91/</td>
<td></td>
<td>True</td>
</tr>
<tr>
<td><code>trace_report_suffix</code></td>
<td>Suffix to add to the trace report filename. Default is the date and time in the format yyyy-MM-dd_HH-mm-ss.</td>
<td><code>string</code></td>
<td></td>
<td></td>
<td>True</td>
</tr>
</tbody>
</table>
<h2 id="reference-genome-options">Reference genome options</h2>
<p>Reference genome related files and options required for the workflow.</p>
<table>
<thead>
<tr>
<th>Parameter</th>
<th>Description</th>
<th>Type</th>
<th>Default</th>
<th>Required</th>
<th>Hidden</th>
</tr>
</thead>
<tbody>
<tr>
<td><code>fasta</code></td>
<td>Reference genome</td>
<td><code>string</code></td>
<td></td>
<td></td>
<td></td>
</tr>
</tbody>
</table>
<h2 id="institutional-config-options">Institutional config options</h2>
<p>Parameters used to describe centralised config profiles. These should not be edited.</p>
<table>
<thead>
<tr>
<th>Parameter</th>
<th>Description</th>
<th>Type</th>
<th>Default</th>
<th>Required</th>
<th>Hidden</th>
</tr>
</thead>
<tbody>
<tr>
<td><code>custom_config_version</code></td>
<td>Git commit id for Institutional configs.</td>
<td><code>string</code></td>
<td>master</td>
<td></td>
<td>True</td>
</tr>
<tr>
<td><code>custom_config_base</code></td>
<td>Base directory for Institutional configs. <details><summary>Help</summary><small>If you're running offline, Nextflow will not be able to fetch the institutional config files from the internet. If you don't need them, then this is not a problem. If you do need them, you should download the files from the repo and tell Nextflow where to find them with this parameter.</small></details></td>
<td><code>string</code></td>
<td>https://raw.githubusercontent.com/nf-core/configs/master</td>
<td></td>
<td>True</td>
</tr>
<tr>
<td><code>config_profile_name</code></td>
<td>Institutional config name.</td>
<td><code>string</code></td>
<td></td>
<td></td>
<td>True</td>
</tr>
<tr>
<td><code>config_profile_description</code></td>
<td>Institutional config description.</td>
<td><code>string</code></td>
<td></td>
<td></td>
<td>True</td>
</tr>
<tr>
<td><code>config_profile_contact</code></td>
<td>Institutional config contact information.</td>
<td><code>string</code></td>
<td></td>
<td></td>
<td>True</td>
</tr>
<tr>
<td><code>config_profile_url</code></td>
<td>Institutional config URL link.</td>
<td><code>string</code></td>
<td></td>
<td></td>
<td>True</td>
</tr>
<tr>
<td><code>publish_unannotated_family_svs</code></td>
<td>Publish unannotated SVs and CNVs per family and caller (these files are only output by default if <code>--skip_sv_annotation</code> or <code>--skip_cnv_calling</code> are active).</td>
<td><code>boolean</code></td>
<td></td>
<td></td>
<td>True</td>
</tr>
</tbody>
</table>
<h2 id="generic-options">Generic options</h2>
<p>Less common options for the pipeline, typically set in a config file.</p>
<table>
<thead>
<tr>
<th>Parameter</th>
<th>Description</th>
<th>Type</th>
<th>Default</th>
<th>Required</th>
<th>Hidden</th>
</tr>
</thead>
<tbody>
<tr>
<td><code>version</code></td>
<td>Display version and exit.</td>
<td><code>boolean</code></td>
<td></td>
<td></td>
<td>True</td>
</tr>
<tr>
<td><code>publish_dir_mode</code></td>
<td>Method used to save pipeline results to output directory. <details><summary>Help</summary><small>The Nextflow <code>publishDir</code> option specifies which intermediate files should be saved to the output directory. This option tells the pipeline what method should be used to move these files. See <a href="https://www.nextflow.io/docs/latest/process.html#publishdir">Nextflow docs</a> for details.</small></details></td>
<td><code>string</code></td>
<td>copy</td>
<td></td>
<td>True</td>
</tr>
<tr>
<td><code>email_on_fail</code></td>
<td>Email address for completion summary, only when pipeline fails. <details><summary>Help</summary><small>An email address to send a summary email to when the pipeline is completed - ONLY sent if the pipeline does not exit successfully.</small></details></td>
<td><code>string</code></td>
<td></td>
<td></td>
<td>True</td>
</tr>
<tr>
<td><code>plaintext_email</code></td>
<td>Send plain-text email instead of HTML.</td>
<td><code>boolean</code></td>
<td></td>
<td></td>
<td>True</td>
</tr>
<tr>
<td><code>max_multiqc_email_size</code></td>
<td>File size limit when attaching MultiQC reports to summary emails.</td>
<td><code>string</code></td>
<td>25.MB</td>
<td></td>
<td>True</td>
</tr>
<tr>
<td><code>monochrome_logs</code></td>
<td>Do not use coloured log outputs.</td>
<td><code>boolean</code></td>
<td></td>
<td></td>
<td>True</td>
</tr>
<tr>
<td><code>hook_url</code></td>
<td>Incoming hook URL for messaging service <details><summary>Help</summary><small>Incoming hook URL for messaging service. Currently, MS Teams and Slack are supported.</small></details></td>
<td><code>string</code></td>
<td></td>
<td></td>
<td>True</td>
</tr>
<tr>
<td><code>multiqc_config</code></td>
<td>Custom config file to supply to MultiQC.</td>
<td><code>string</code></td>
<td></td>
<td></td>
<td>True</td>
</tr>
<tr>
<td><code>multiqc_logo</code></td>
<td>Custom logo file to supply to MultiQC. File name must also be set in the MultiQC config file</td>
<td><code>string</code></td>
<td></td>
<td></td>
<td>True</td>
</tr>
<tr>
<td><code>multiqc_methods_description</code></td>
<td>Custom MultiQC yaml file containing HTML including a methods description.</td>
<td><code>string</code></td>
<td></td>
<td></td>
<td></td>
</tr>
<tr>
<td><code>validate_params</code></td>
<td>Boolean whether to validate parameters against the schema at runtime</td>
<td><code>boolean</code></td>
<td>True</td>
<td></td>
<td>True</td>
</tr>
</tbody>
</table>
<h2 id="workflow-options">Workflow options</h2>
<p>Workflow options specific to genomic-medicine-sweden/nallo</p>
<table>
<thead>
<tr>
<th>Parameter</th>
<th>Description</th>
<th>Type</th>
<th>Default</th>
<th>Required</th>
<th>Hidden</th>
</tr>
</thead>
<tbody>
<tr>
<td><code>preset</code></td>
<td>Enable or disable certain parts of the pipeline by default, depending on data type (<code>revio</code>, <code>pacbio</code>, <code>ONT_R10</code>)</td>
<td><code>string</code></td>
<td>revio</td>
<td>True</td>
<td></td>
</tr>
<tr>
<td><code>snv_caller</code></td>
<td>Which short variant software to use (<code>deepvariant</code>)</td>
<td><code>string</code></td>
<td>deepvariant</td>
<td></td>
<td></td>
</tr>
<tr>
<td><code>sv_caller</code></td>
<td>From which SV caller to merge (with CNVs), annotate, rank and filter variants from (<code>severus</code> or <code>sniffles</code>).</td>
<td><code>string</code></td>
<td>severus</td>
<td></td>
<td></td>
</tr>
<tr>
<td><code>str_caller</code></td>
<td>Which caller to use for short tandem repeat expansions (TRGT or STRdust).</td>
<td><code>string</code></td>
<td>trgt</td>
<td></td>
<td></td>
</tr>
<tr>
<td><code>phaser</code></td>
<td>Which phasing software to use (<code>longphase</code>, <code>whatshap</code>, <code>hiphase</code>)</td>
<td><code>string</code></td>
<td>longphase</td>
<td></td>
<td></td>
</tr>
<tr>
<td><code>hifiasm_mode</code></td>
<td>Run hifiasm in hifi-only or hifi-trio mode (<code>hifi-only</code>, <code>trio-binning</code>)</td>
<td><code>string</code></td>
<td>hifi-only</td>
<td></td>
<td></td>
</tr>
<tr>
<td><code>hifiasm_preset</code></td>
<td>Hifiasm preset, is set to <code>--ont</code> when <code>--profile ONT_R10</code> is active.</td>
<td><code>string</code></td>
<td></td>
<td></td>
<td></td>
</tr>
<tr>
<td><code>alignment_processes</code></td>
<td>If alignment_processes is bigger than 1, input files will be split and aligned in parallel to reduce processing time.</td>
<td><code>integer</code></td>
<td>8</td>
<td></td>
<td></td>
</tr>
<tr>
<td><code>snv_calling_processes</code></td>
<td>If snv_calling_processes is bigger than 1, short variant calling will be done in parallel to reduce processing time.</td>
<td><code>integer</code></td>
<td>13</td>
<td></td>
<td></td>
</tr>
<tr>
<td><code>vep_cache_version</code></td>
<td>VEP cache version</td>
<td><code>integer</code></td>
<td>110</td>
<td></td>
<td></td>
</tr>
<tr>
<td><code>vep_plugin_files</code></td>
<td>A csv file with vep_files as header, and then paths to vep plugin files. Paths to pLI_values.txt and LoFtool_scores.txt are required.</td>
<td><code>string</code></td>
<td></td>
<td></td>
<td></td>
</tr>
<tr>
<td><code>filter_variants_hgnc_ids</code></td>
<td>A tsv/csv file with a <code>hgnc_ids</code> column header, and then one numerical HGNC ID per row. E.g. <code>4281</code> or <code>HGNC:4281</code>.</td>
<td><code>string</code></td>
<td></td>
<td></td>
<td></td>
</tr>
<tr>
<td><code>filter_snvs_expression</code></td>
<td>An expression that is passed to bcftools view to filter SNVs, e.g. --filter_snvs_expression "-e 'INFO/AQ&gt;60'"</td>
<td><code>string</code></td>
<td></td>
<td></td>
<td></td>
</tr>
<tr>
<td><code>filter_svs_expression</code></td>
<td>An expression that is passed to bcftools view to filter SVs, e.g. --filter_svs_expression "-e 'INFO/AQ&gt;60'"</td>
<td><code>string</code></td>
<td></td>
<td></td>
<td></td>
</tr>
<tr>
<td><code>deepvariant_model_type</code></td>
<td>Sets the model type used for DeepVariant. This is set automatically using <code>--preset</code> by default.</td>
<td><code>string</code></td>
<td>PACBIO</td>
<td></td>
<td>True</td>
</tr>
<tr>
<td><code>minimap2_read_mapping_preset</code></td>
<td>Sets the minimap2-preset (-x) for read alignment. This is set automatically using the pipeline <code>--preset</code> by default.</td>
<td><code>string</code></td>
<td>map-hifi</td>
<td></td>
<td>True</td>
</tr>
<tr>
<td><code>extra_modkit_options</code></td>
<td>Extra options to modkit, used for test profile.</td>
<td><code>string</code></td>
<td></td>
<td></td>
<td>True</td>
</tr>
<tr>
<td><code>extra_vep_options</code></td>
<td>Extra options to VEP, used for test profile.</td>
<td><code>string</code></td>
<td></td>
<td></td>
<td>True</td>
</tr>
<tr>
<td><code>extra_paraphase_options</code></td>
<td>Extra options to Paraphase, used for test profile.</td>
<td><code>string</code></td>
<td></td>
<td></td>
<td>True</td>
</tr>
<tr>
<td><code>extra_hifiasm_options</code></td>
<td>Extra options to hifiasm, used for test profile.</td>
<td><code>string</code></td>
<td></td>
<td></td>
<td>True</td>
</tr>
</tbody>
</table>












                
              </article>
            </div>
          
          
<script>var target=document.getElementById(location.hash.slice(1));target&&target.name&&(target.checked=target.name.startsWith("__tabbed_"))</script>
        </div>
        
          <button type="button" class="md-top md-icon" data-md-component="top" hidden>
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M13 20h-2V8l-5.5 5.5-1.42-1.42L12 4.16l7.92 7.92-1.42 1.42L13 8z"/></svg>
  Back to top
</button>
        
      </main>
      
        <footer class="md-footer">
  
    
      
      <nav class="md-footer__inner md-grid" aria-label="Footer" >
        
          
          <a href="../output/" class="md-footer__link md-footer__link--prev" aria-label="Previous: Output">
            <div class="md-footer__button md-icon">
              
              <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11z"/></svg>
            </div>
            <div class="md-footer__title">
              <span class="md-footer__direction">
                Previous
              </span>
              <div class="md-ellipsis">
                Output
              </div>
            </div>
          </a>
        
        
          
          <a href="../CITATIONS/" class="md-footer__link md-footer__link--next" aria-label="Next: Citations">
            <div class="md-footer__title">
              <span class="md-footer__direction">
                Next
              </span>
              <div class="md-ellipsis">
                Citations
              </div>
            </div>
            <div class="md-footer__button md-icon">
              
              <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M4 11v2h12l-5.5 5.5 1.42 1.42L19.84 12l-7.92-7.92L10.5 5.5 16 11z"/></svg>
            </div>
          </a>
        
      </nav>
    
  
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-copyright">
  
  
    Made with
    <a href="https://squidfunk.github.io/mkdocs-material/" target="_blank" rel="noopener">
      Material for MkDocs
    </a>
  
</div>
      
    </div>
  </div>
</footer>
      
    </div>
    <div class="md-dialog" data-md-component="dialog">
      <div class="md-dialog__inner md-typeset"></div>
    </div>
    
    
    
      
      <script id="__config" type="application/json">{"base": "..", "features": ["content.code.annotate", "content.code.copy", "navigation.instant", "navigation.top", "navigation.tracking", "navigation.sections", "navigation.footer", "search.share", "toc.follow"], "search": "../assets/javascripts/workers/search.d50fe291.min.js", "tags": null, "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version": "Select version"}, "version": {"provider": "mike"}}</script>
    
    
      <script src="../assets/javascripts/bundle.13a4f30d.min.js"></script>
      
    
  </body>
</html>