Changes in module 'nf-core/bcftools/annotate'
--- modules/nf-core/bcftools/annotate/meta.yml
+++ modules/nf-core/bcftools/annotate/meta.yml
@@ -34,6 +34,9 @@
   - header_lines:
       type: file
       description: Contains lines to append to the output VCF header
+  - rename_chrs:
+      type: file
+      description: Contains new and old chromosome names
 output:
   - meta:
       type: map

--- modules/nf-core/bcftools/annotate/main.nf
+++ modules/nf-core/bcftools/annotate/main.nf
@@ -8,10 +8,9 @@
         'biocontainers/bcftools:1.20--h8b25389_0' }"
 
     input:
-    tuple val(meta), path(input), path(index)
-    path(annotations)
-    path(annotations_index)
+    tuple val(meta), path(input), path(index), path(annotations), path(annotations_index)
     path(header_lines)
+    path(rename_chrs)
 
     output:
     tuple val(meta), path("*.{vcf,vcf.gz,bcf,bcf.gz}"), emit: vcf
@@ -27,6 +26,7 @@
     def prefix  = task.ext.prefix ?: "${meta.id}"
     def header_file = header_lines ? "--header-lines ${header_lines}" : ''
     def annotations_file = annotations ? "--annotations ${annotations}" : ''
+    def rename_chrs_file = rename_chrs ? "--rename-chrs ${rename_chrs}" : ''
     def extension = args.contains("--output-type b") || args.contains("-Ob") ? "bcf.gz" :
                     args.contains("--output-type u") || args.contains("-Ou") ? "bcf" :
                     args.contains("--output-type z") || args.contains("-Oz") ? "vcf.gz" :
@@ -43,6 +43,7 @@
         $args \\
         $annotations_file \\
         $header_file \\
+        $rename_chrs_file \\
         --output ${prefix}.${extension} \\
         --threads $task.cpus \\
         $input

************************************************************
