Changes in module 'nf-core/gfastats'
--- modules/nf-core/gfastats/meta.yml
+++ modules/nf-core/gfastats/meta.yml
@@ -30,7 +30,7 @@
       type: string
       description: Output format (fasta, fastq, gfa)
   - genome_size:
-      type: number
+      type: float
       description: estimated genome size (bp) for NG* statistics (optional).
   - target:
       type: string

--- modules/nf-core/gfastats/main.nf
+++ modules/nf-core/gfastats/main.nf
@@ -1,6 +1,6 @@
 process GFASTATS {
     tag "$meta.id"
-    label 'process_low'
+    label 'process_single'
 
     conda "bioconda::gfastats=1.3.5"
     container "${ workflow.containerEngine == 'singularity' && !task.ext.singularity_pull_docker_container ?
@@ -40,11 +40,11 @@
         $ibed \\
         $ebed \\
         $sak \\
-        --out-format ${prefix}.${out_fmt}.gz \\
+        --out-format ${assembly.baseName}.${out_fmt}.gz \\
         $assembly \\
         $genome_size \\
         $target \\
-        > ${prefix}.assembly_summary
+        > ${assembly.baseName}.assembly_summary
 
     cat <<-END_VERSIONS > versions.yml
     "${task.process}":

************************************************************
