Changes in module 'nf-core/hifiasm'
--- modules/nf-core/hifiasm/main.nf
+++ modules/nf-core/hifiasm/main.nf
@@ -8,9 +8,7 @@
         'biocontainers/hifiasm:0.19.8--h43eeafb_0' }"
 
     input:
-    tuple val(meta), path(reads)
-    path  paternal_kmer_dump
-    path  maternal_kmer_dump
+    tuple val(meta), path(reads, stageAs: "?/*"), path(paternal_kmer_dump), path(maternal_kmer_dump)
     path  hic_read1
     path  hic_read2
 

************************************************************
