Changes in module 'nf-core/hifiasm'
--- modules/nf-core/hifiasm/main.nf
+++ modules/nf-core/hifiasm/main.nf
@@ -2,10 +2,10 @@
     tag "$meta.id"
     label 'process_high'
 
-    conda "bioconda::hifiasm=0.18.5"
+    conda "bioconda::hifiasm=0.19.5"
     container "${ workflow.containerEngine == 'singularity' && !task.ext.singularity_pull_docker_container ?
-        'https://depot.galaxyproject.org/singularity/hifiasm:0.18.5--h5b5514e_0' :
-        'biocontainers/hifiasm:0.18.5--h5b5514e_0' }"
+        'https://depot.galaxyproject.org/singularity/hifiasm:0.19.5--h5b5514e_0' :
+        'quay.io/biocontainers/hifiasm:0.19.5--h5b5514e_0' }"
 
     input:
     tuple val(meta), path(reads)

************************************************************
