Changes in module 'nf-core/minimap2/index'
--- modules/nf-core/minimap2/index/main.nf
+++ modules/nf-core/minimap2/index/main.nf
@@ -1,11 +1,11 @@
 process MINIMAP2_INDEX {
-    label 'process_medium'
+    label 'process_low'
 
     // Note: the versions here need to match the versions used in minimap2/align
-    conda "bioconda::minimap2=2.24"
+    conda "bioconda::minimap2=2.26"
     container "${ workflow.containerEngine == 'singularity' && !task.ext.singularity_pull_docker_container ?
-        'https://depot.galaxyproject.org/singularity/minimap2:2.24--h7132678_1' :
-        'biocontainers/minimap2:2.24--h7132678_1' }"
+        'https://depot.galaxyproject.org/singularity/minimap2:2.26--h7132678_0' :
+        'biocontainers/minimap2:2.26--h7132678_0' }"
 
     input:
     tuple val(meta), path(fasta)

************************************************************
